{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Intermediate Bioinformatics for Parkinson’s Disease Genetics\n",
    "\n",
    "\n",
    " - **Module II:** Investigating the cumulative effect of genetic variation on Parkinson’s disease (Haplotypes and Rare variant analyses)\n",
    " \n",
    " - **Author(s):** Kajsa Brolin on behalf of the Global Parkinson's Genetics Program (GP2) from Aligning Science Across Parkinson's (ASAP)\n",
    " \n",
    " - **Estimated Computation and Runtime:**\n",
    "   - Estimated Specifications: 4 CPUs/15 GB, persistent disk size 300 GB\n",
    "   - Estimated Runtime:  8.5 h total. The identification of haplotype blocks for the different populations take the most time, ~7h. Output files are uploaded that can be used to reduce runtime to only 1.5h). \n",
    "   \n",
    " - **Application configuration:** Legacy R/Bioconductor (R 4.1.1, Bioconductor 3.13, Python 3.7.10)\n",
    " \n",
    " - **Date Last Updated:** 15-AUGUST-2022\n",
    "     - Update Description: The --kernel flag was removed from the pathway-based analysis. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quick Description:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook sets up to identify haplotype blocks and haplotypes in your data set and to analyse their associaton to PD. As an axample, this notebook specifically focus on haplotypes in SNCA. This notebook also sets up to run gene-based rare variant analyses (also with SNCA as an example) as well as pathway-based burden analyses using two lysosomal pathways as an example."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background/Motivation:\n",
    "\n",
    "Haplotype analyses have been crucial to detect genetic risk factors in PD. They vary across populations and this will be key for further fine-mapping analyses since the variation of haplotypes across populations can help further narrow down loci of interest, or even unravel rare variants linked to PD. Identification of rare variants associated with PD is difficult for several reasons (e.g. very large sample sizes are needed!). Given the limitations for studying rare variants, there are current methods focused on exploring the cumulative effect of multiple rare variants at a gene, multi-gene or region level. Examples of these tests are burden tests and the non-burden sequence kernel association test that we have covered in this course.\n",
    "\n",
    "!NOTE!\n",
    "Information from https://www.cog-genomics.org/plink/1.9/ld regarding haplotypes in PLINK:\n",
    "The .blocks file is valid input for PLINK 1.07's --hap command. However, the --hap... family of flags has not been reimplemented in PLINK 1.9 due to poor phasing accuracy (and, consequently, inferior haplotype likelihood/frequency estimates) relative to other software; for now, we recommend using BEAGLE 3.3.2 instead of PLINK for case/control haplotype association analysis. \n",
    "\n",
    "Here, we will look at som examples of haplotype and rare variant analyses"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow summary"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "  0. Get started - Set up environment, download needed softwares, and import files\n",
    "  \n",
    "**Haplotype analyses**\n",
    "  1. Identifying haplotype blocks and haplotypes in your dataset\n",
    "  2. Analyzing associations between identified haplotypes in SNCA and PD risk in your dataset\n",
    "  3. Comparing the size of haplotype blocks at PD risk loci across different populations\n",
    "  \n",
    "**Rare variant analyses**\n",
    "\n",
    "  4. Gene/Region-based analysis (including annotation using ANNOVAR)\n",
    "  5. Pathway-based analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Workflow"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [0. Get started - Set up environment, download needed softwares, and import files](#0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section goes through:\n",
    "* Setting up Python libraries, data path variables, billing variables, and functions\n",
    "* Setting up billing variables\n",
    "* Installing PLINK v1.9\n",
    "* Install rpy2 to run R in python\n",
    "* Install RVTESTS\n",
    "* Install ANNOVAR\n",
    "* Download files that are needed in ANNOVAR\n",
    "* Copy the needed data/files to your workspace"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [1. Haplotype analyses - Identifying haplotype blocks and haplotypes in your dataset](#1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section goes through:\n",
    "* How to extract the region of interest\n",
    "* Calculate haplotype blocks"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [2. Haplotype analyses - Analyzing associations between identified haplotypes in SNCA and PD risk in your dataset](#2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section goes through:\n",
    "* Recode your data to ped and map format\n",
    "* Set up libraries and load packages for R analyses\n",
    "* Add SNP name to the PED file\n",
    "* Create haplotype files\n",
    "* Run association analyses (both non-adjusted and adjusted) for each haplotype"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [3. Haplotype analyses - Comparing the size of haplotype blocks at PD risk loci across different populations](#3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section goes through:\n",
    "* Extract cases from both the 1000G YRI and ACB populations\n",
    "* Extract common SNPs between both datasets\n",
    "* Calculate haplotype blocks in both cohorts (This step takes quite long time)\n",
    "* Compare the length and number of SNPs of haplotypes at the 92 risk loci between the two cohorts"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [4. Rare variant analyses - Gene/Region-based analysis](#4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section goes through:\n",
    "* How to extract the region of interest (again...different data though!)\n",
    "* Recode the PLINK files to vcf (+bgzip and tabix the vcf file)\n",
    "* Annotate the vcf file using ANNOVAR\n",
    "* Clean up the output file, generate a list of non-coding variants and extract the them from the data\n",
    "* Recode to generate a .ped file with phenotype info to use in RVTESTS\n",
    "* Bgzip, tabix and run rare variant analyses in RVTESTS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [5. Rare variant analyses - Pathway-based analysis](#5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section goes through:\n",
    "* Run gene-set based pathway test in RVESTs\n",
    "* Recode the PLINK format file to vcf and bgzip\n",
    "* Run the pathway-based burden analysis in RVTESTS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 0. Getting started\n",
    "<a id=\"0\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set up cells copied from the GP2 Beginners Bioinformatics course, https://github.com/GP2-TNC-WG/GP2-Bioinformatics-course/blob/master/Introduction.md"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the os package to interact with the environment\n",
    "import os\n",
    "\n",
    "# Bring in Pandas for Dataframe functionality\n",
    "import pandas as pd\n",
    "\n",
    "# numpy for basics\n",
    "import numpy as np\n",
    "\n",
    "# Use StringIO for working with file contents\n",
    "from io import StringIO\n",
    "\n",
    "# Enable IPython to display matplotlib graphs\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline\n",
    "\n",
    "# Enable interaction with the FireCloud API\n",
    "from firecloud import api as fapi\n",
    "\n",
    "# Import the iPython HTML rendering for displaying links to Google Cloud Console\n",
    "from IPython.core.display import display, HTML\n",
    "\n",
    "# Import urllib modules for building URLs to Google Cloud Console\n",
    "import urllib.parse\n",
    "\n",
    "# BigQuery for querying data\n",
    "from google.cloud import bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set up billing project and data path variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "terra-9b559320\n"
     ]
    }
   ],
   "source": [
    "BILLING_PROJECT_ID = os.environ['GOOGLE_PROJECT']\n",
    "WORKSPACE_NAMESPACE = os.environ['WORKSPACE_NAMESPACE']\n",
    "WORKSPACE_NAME = os.environ['WORKSPACE_NAME']\n",
    "WORKSPACE_BUCKET = os.environ['WORKSPACE_BUCKET']\n",
    "\n",
    "WORKSPACE_ATTRIBUTES = fapi.get_workspace(WORKSPACE_NAMESPACE, WORKSPACE_NAME).json().get('workspace',{}).get('attributes',{})\n",
    "\n",
    "print(BILLING_PROJECT_ID)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set up useful functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Utility routine for printing a shell command before executing it\n",
    "def shell_do(command):\n",
    "    print(f'Executing: {command}')\n",
    "    !$command\n",
    "    \n",
    "def gcs_read_csv(path, sep=None):\n",
    "    \"\"\"Return a DataFrame from the contents of a delimited file in GCS\"\"\"\n",
    "    return pd.read_csv(StringIO(gcs_read_file(path)), sep=sep, engine='python')\n",
    "\n",
    "# Utility routine for display a message and a link\n",
    "def display_html_link(description, link_text, url):\n",
    "    html = f'''\n",
    "    <p>\n",
    "    </p>\n",
    "    <p>\n",
    "    {description}\n",
    "    <a target=_blank href=\"{url}\">{link_text}</a>.\n",
    "    </p>\n",
    "    '''\n",
    "\n",
    "    display(HTML(html))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Haplotype analyses:**\n",
    "* Programs that you need:\n",
    "    - PLINK v1.9\n",
    "    - R\n",
    "* Files that you will need:\n",
    "    - QC:ed pre-imputation PLINK binary files containing cases and controls (.fam, .bim, .bed files)\n",
    "    - Covariate file: covariateFile_Haplo.txt\n",
    "    - QC:ed pre-imputation PLINK binary files containing cases of different ancestries (.fam, .bim, .bed files)\n",
    "    - PD loci file, txt file containing SNP, CHR, and BP for 92 reported PD risk loci: 92SNPs_haplotype.txt\n",
    "    \n",
    "**Rare variant analyses:**\n",
    "* Programs that you need:\n",
    "    - PLINK v1.9\n",
    "    - RVTESTS\n",
    "    - ANNOVAR\n",
    "* Files that you will need:\n",
    "    - Imputed QC:ed PLINK binary files with case control status (IMPUTED.HARDCALLS.Demo_formatted.bed, bim and fam files). These have been filtered for MAF > 0.001 and Rsq (imputation quality score) > 0.8 (Genome build GRCh37)\n",
    "    - RefFlat file for RVTESTS: refFlat_hg19.txt. Can be downloaded from: http://qbrc.swmed.edu/zhanxw/seqminer/data/refFlat_hg19.txt.gz (for GRCh37)\n",
    "    - Covariate file (we will use the same as for the haplotype analyses): covariateFile_Haplo.txt\n",
    "    - SetFile: Needed to run pathway based burden analysis. Each pathway has it's own line, and gene coordinates follow in this format (no spaces between commas) 16:68245303-68261058,15:82659280-82709946"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Install PLINK v1.9**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%bash\n",
    "\n",
    "if test -e /home/jupyter/plink; then\n",
    "\n",
    "echo \"Plink is already installed in /home/jupyter/\"\n",
    "else\n",
    "echo \"Plink is not installed\"\n",
    "cd /home/jupyter\n",
    "\n",
    "wget http://s3.amazonaws.com/plink1-assets/plink_linux_x86_64_20190304.zip \n",
    "\n",
    "unzip -o plink_linux_x86_64_20190304.zip\n",
    "\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Install rpy2 to run R in python**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: rpy2 in /home/jupyter/.local/lib/python3.7/site-packages (3.5.3)\n",
      "Requirement already satisfied: tzlocal in /home/jupyter/.local/lib/python3.7/site-packages (from rpy2) (4.2)\n",
      "Requirement already satisfied: pytz in /opt/conda/lib/python3.7/site-packages (from rpy2) (2021.3)\n",
      "Requirement already satisfied: jinja2 in /opt/conda/lib/python3.7/site-packages (from rpy2) (2.11.3)\n",
      "Requirement already satisfied: cffi>=1.10.0 in /opt/conda/lib/python3.7/site-packages (from rpy2) (1.14.6)\n",
      "Requirement already satisfied: typing-extensions in /opt/conda/lib/python3.7/site-packages (from rpy2) (3.10.0.2)\n",
      "Requirement already satisfied: pycparser in /opt/conda/lib/python3.7/site-packages (from cffi>=1.10.0->rpy2) (2.20)\n",
      "Requirement already satisfied: MarkupSafe>=0.23 in /opt/conda/lib/python3.7/site-packages (from jinja2->rpy2) (1.1.1)\n",
      "Requirement already satisfied: backports.zoneinfo in /home/jupyter/.local/lib/python3.7/site-packages (from tzlocal->rpy2) (0.2.1)\n",
      "Requirement already satisfied: pytz-deprecation-shim in /home/jupyter/.local/lib/python3.7/site-packages (from tzlocal->rpy2) (0.1.0.post0)\n",
      "Requirement already satisfied: tzdata in /home/jupyter/.local/lib/python3.7/site-packages (from pytz-deprecation-shim->tzlocal->rpy2) (2022.2)\n",
      "\u001b[33mWARNING: You are using pip version 21.3.1; however, version 22.2.2 is available.\n",
      "You should consider upgrading via the '/opt/conda/bin/python3.7 -m pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install rpy2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import rpy2\n",
    "import rpy2.rinterface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Install RVTESTS**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rvtests is not installed\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Cloning into 'rvtests'...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "if test -e ~/home/jupyter/rvtests; then\n",
    "\n",
    "echo \"rvtests is already installed\"\n",
    "else\n",
    "echo \"rvtests is not installed\"\n",
    "cd /home/jupyter\n",
    "\n",
    "git clone https://github.com/zhanxw/rvtests\n",
    "\n",
    "fi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%bash\n",
    "\n",
    "cd /home/jupyter/rvtests\n",
    "make"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Install ANNOVAR**\n",
    "\n",
    "Important! You need to add the download link after registration on the annovar website: https://www.openbioinformatics.org/annovar/annovar_download_form.php"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%bash\n",
    "\n",
    "if test -e /home/jupyter/annovar; then\n",
    "\n",
    "echo \"annovar is already installed in /home/jupyter\"\n",
    "else\n",
    "echo \"annovar is not installed\"\n",
    "cd /home/jupyter/\n",
    "\n",
    "wget http://www.openbioinformatics.org/annovar/download/0wgxR2rIVP/annovar.latest.tar.gz\n",
    "\n",
    "tar xvfz annovar.latest.tar.gz\n",
    "\n",
    "fi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Download files that are needed in ANNOVAR:\n",
    "- These files are quite large and take some timw to download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%bash\n",
    "cd /home/jupyter/annovar/\n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar refGene humandb/\n",
    "perl annotate_variation.pl -buildver hg19 -downdb cytoBand humandb/\n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar ensGene humandb/\n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar exac03 humandb/ \n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar avsnp147 humandb/ \n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar dbnsfp30a humandb/\n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar gnomad211_genome humandb/\n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar ljb26_all humandb/\n",
    "perl annotate_variation.pl -buildver hg19 -downdb -webfrom annovar clinvar_20140902 humandb/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Check that you can see all the install programs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annovar\n",
      "annovar.latest.tar.gz\n",
      "GP2 Bioinformatics Course 2\n",
      "jupyter.log\n",
      "LICENSE\n",
      "lost+found\n",
      "packages\n",
      "plink\n",
      "plink_linux_x86_64_20190304.zip\n",
      "prettify\n",
      "rvtests\n",
      "toy.map\n",
      "toy.ped\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/jupyter/\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Copy the needed data/files to your workspace**\n",
    "\n",
    "General tip: save data to your Persistent Disk (either /home/jupyter-user/ or /home/jupyter/) - you can create directories within this. That way you can delete and recreate your Cloud Environment if there are any issues, and the data saved on the Persistent Disk will be retained. See: https://support.terra.bio/hc/en-us/articles/360047318551-Detachable-Persistent-Disks-\n",
    "\n",
    "The files needed for the tutorial are found in this module. They need to be uploaded manually to your bucket before importing them to the workspace.\n",
    "\n",
    "The files have previously undergone QC and been randomly assigned case/control status. Data in the covariate file is has also been generated randomly and do not represent true data. We will import all files needed for the different haplotype analyses at this step."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Make directions for the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /home/jupyter/1000G_TEST/\n",
    "mkdir -p /home/jupyter/Burden/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Copy files into your workspace > Data > Files\n",
    "* Import files from the workspace bucket into the workspace"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/FILTERED.test_formatted.* /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/FILTERED.test_formatted.bim...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/FILTERED.test_formatted.bed...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/FILTERED.test_formatted.fam...\n",
      "\\ [3/3 files][108.0 MiB/108.0 MiB] 100% Done                                    \n",
      "Operation completed over 3 objects/108.0 MiB.                                    \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/covariateFile_Haplo.txt /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/covariateFile_Haplo.txt...\n",
      "/ [1/1 files][ 55.9 KiB/ 55.9 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/55.9 KiB.                                     \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_ALL_1000.* /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_ALL_1000.bed...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_ALL_1000.bim...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_ALL_1000.fam...\n",
      "/ [3/3 files][  1.7 GiB/  1.7 GiB] 100% Done  37.8 MiB/s ETA 00:00:00           \n",
      "Operation completed over 3 objects/1.7 GiB.                                      \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_ALL_1000g.* /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_ALL_1000g.bed...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_ALL_1000g.bim...       \n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_ALL_1000g.fam...       \n",
      "/ [3/3 files][  1.8 GiB/  1.8 GiB] 100% Done  34.2 MiB/s ETA 00:00:00           \n",
      "Operation completed over 3 objects/1.8 GiB.                                      \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/92SNPs_haplotype.txt /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/92SNPs_haplotype.txt...\n",
      "/ [1/1 files][  2.0 KiB/  2.0 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/2.0 KiB.                                      \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/IMPUTED.HARDCALLS.Demo_formatted.* /home/jupyter/Burden/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/IMPUTED.HARDCALLS.Demo_formatted.bed...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/IMPUTED.HARDCALLS.Demo_formatted.bim...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/IMPUTED.HARDCALLS.Demo_formatted.fam...\n",
      "\\ [3/3 files][  1.5 GiB/  1.5 GiB] 100% Done  36.7 MiB/s ETA 00:00:00           \n",
      "Operation completed over 3 objects/1.5 GiB.                                      \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/refFlat_hg19.txt /home/jupyter/Burden/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/refFlat_hg19.txt...\n",
      "/ [1/1 files][  9.2 MiB/  9.2 MiB] 100% Done                                    \n",
      "Operation completed over 1 objects/9.2 MiB.                                      \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/covariateFile_Haplo.txt /home/jupyter/Burden/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/covariateFile_Haplo.txt...\n",
      "/ [1/1 files][ 55.9 KiB/ 55.9 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/55.9 KiB.                                     \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/lysosomal-setfile-example.txt /home/jupyter/Burden/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/lysosomal-setfile-example.txt...\n",
      "/ [1/1 files][  3.2 KiB/  3.2 KiB] 100% Done                                    \n",
      "Operation completed over 1 objects/3.2 KiB.                                      \n"
     ]
    }
   ],
   "source": [
    "#HAPLOTYPE ANALYSES\n",
    "#QC:ed case/control files (non-imputed)\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/FILTERED.test_formatted.* /home/jupyter/1000G_TEST/')\n",
    "#Covariate file\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/covariateFile_Haplo.txt /home/jupyter/1000G_TEST/')\n",
    "#QC:ed case files (non-imputed) with different ancestry (ACB population)\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/ACB_ALL_1000.* /home/jupyter/1000G_TEST/')\n",
    "#QC:ed case files (non-imputed) with different ancestry (YRI population)\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/YRI_ALL_1000g.* /home/jupyter/1000G_TEST/')\n",
    "#PD risk loci file:\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/92SNPs_haplotype.txt /home/jupyter/1000G_TEST/')\n",
    "\n",
    "#RARE VARIANT ANALYSES\n",
    "#Imputed QC:ed PLINK binary files with case control status (IMPUTED.HARDCALLS.Demo_formatted)\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/IMPUTED.HARDCALLS.Demo_formatted.* /home/jupyter/Burden/')\n",
    "#RefFlat txt file:\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/refFlat_hg19.txt /home/jupyter/Burden/')\n",
    "#Covariate file\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/covariateFile_Haplo.txt /home/jupyter/Burden/')\n",
    "#SetFile\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/lysosomal-setfile-example.txt /home/jupyter/Burden/')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Check that you can see the data in the folders:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "92SNPs_haplotype.txt\n",
      "ACB_ALL_1000.bed\n",
      "ACB_ALL_1000.bim\n",
      "ACB_ALL_1000.fam\n",
      "covariateFile_Haplo.txt\n",
      "FILTERED.test_formatted.bed\n",
      "FILTERED.test_formatted.bim\n",
      "FILTERED.test_formatted.fam\n",
      "YRI_ALL_1000g.bed\n",
      "YRI_ALL_1000g.bim\n",
      "YRI_ALL_1000g.fam\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ls /home/jupyter/1000G_TEST/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "covariateFile_Haplo.txt\n",
      "IMPUTED.HARDCALLS.Demo_formatted.bed\n",
      "IMPUTED.HARDCALLS.Demo_formatted.bim\n",
      "IMPUTED.HARDCALLS.Demo_formatted.fam\n",
      "lysosomal-setfile-example.txt\n",
      "refFlat_hg19.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "ls /home/jupyter/Burden/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Haplotype analysis - Identifying haplotype blocks and haplotypes in your dataset\n",
    "<a id=\"1\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* For this, we will focus on a specific genomic region and will the region of the gene SNCA as an example\n",
    "\n",
    "* SNCA position on GRCh37/hg19:chr4:90645250-90759466 (Ensembl)\n",
    "\n",
    "* First, we create a new folder for the analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /home/jupyter/1000G_TEST/SNCA_hap"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Extract the region of interest**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/FILTERED.test_formatted --chr 4 --from-bp 90645250 --to-bp 90759466 --maf 0.01 --make-bed --out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/1000G_TEST/FILTERED.test_formatted\n",
      "  --chr 4\n",
      "  --from-bp 90645250\n",
      "  --maf 0.01\n",
      "  --make-bed\n",
      "  --out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA\n",
      "  --to-bp 90759466\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "43 out of 869296 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "3 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "40 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--make-bed to\n",
      "/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.bed +\n",
      "/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.bim +\n",
      "/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/FILTERED.test_formatted \\\n",
    "--chr 4 \\\n",
    "--from-bp 90645250 \\\n",
    "--to-bp 90759466 \\\n",
    "--maf 0.01 \\\n",
    "--make-bed \\\n",
    "--out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Calculate haplotype blocks**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The flag --blocks in PLINK estimates haplotype blocks via Haploview's interpretation of the block definition suggested by Gabriel S et al. (2002) The Structure of Haplotype Blocks in the Human Genome. Each block's variant IDs are written to plink.blocks, and a longer report with position information is written to plink.blocks.det."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA --maf 0.01 --blocks --out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA_blocks\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA_blocks.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA\n",
      "  --blocks\n",
      "  --maf 0.01\n",
      "  --out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA_blocks\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "40 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "0 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "40 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--blocks: 4 haploblocks written to505356606366707376808386909396%\n",
      "/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA_blocks.blocks .\n",
      "Extra block details written to\n",
      "/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA_blocks.blocks.det\n",
      ".\n",
      "Longest span: 41.128kb.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA \\\n",
    "--maf 0.01 \\\n",
    "--blocks \\\n",
    "--out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA_blocks')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Two files will be generated (besides the log file):\n",
    "* FILTERED.test_formatted.SNCA_blocks.blocks\n",
    "* FILTERED.test_formatted.SNCA_blocks.blocks.det\n",
    "\n",
    ".blocks:\n",
    "The identified blocks. Contain one line per block, each with an asterisk followed by variant IDs, i.e.,\n",
    "* 4:90645674 4:90653134 4:90663542 4:90663670 4:90668019\n",
    "\n",
    ".block.det:\n",
    "A more detailed file about the blocks. Have a header line, followed by one line per block with the following six fields:\n",
    "\n",
    "* CHR\tChromosome code\n",
    "* BP1\tFirst base-pair coordinate\n",
    "* BP2\tLast base-pair coordinate\n",
    "* KB\tBlock length in kbs\n",
    "* NSNPS Number of variants in block\n",
    "* SNPS '|'-delimited variant IDs\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Haplotype analysis - Analyzing associations between identified haplotypes in SNCA and PD risk in your dataset\n",
    "<a id=\"2\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Recode your data to ped and map format (we need those files for the analyses in R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA --recode --out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA\n",
      "  --out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA\n",
      "  --recode\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "40 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "40 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--recode ped to\n",
      "/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.ped +\n",
      "/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.map ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA \\\n",
    "--recode \\\n",
    "--out /home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Set up libraries for R analyses:\n",
    "- Make directory R_packages if it does not already exist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /home/jupyter/R_packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: also installing the dependencies ‘glue’, ‘cli’, ‘rlang’, ‘vctrs’\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/glue_1.6.2.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 106510 bytes (104 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 104 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/cli_3.3.0.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 495823 bytes (484 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 484 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/rlang_1.0.4.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 740971 bytes (723 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 723 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/vctrs_0.4.1.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 916782 bytes (895 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 895 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/dplyr_1.0.9.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 957637 bytes (935 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 935 KB\n",
      "\n",
      "\n",
      "* installing *source* package ‘glue’ ...\n",
      "** package ‘glue’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c glue.c -o glue.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c trim.c -o trim.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o glue.so glue.o init.o trim.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-glue/00new/glue/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (glue)\n",
      "* installing *source* package ‘rlang’ ...\n",
      "** package ‘rlang’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang/    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c capture.c -o capture.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang/    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c internal.c -o internal.o\n",
      "g++ -std=gnu++11 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang/     -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rlang-rcc.cpp -o rlang-rcc.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang/    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rlang.c -o rlang.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang/    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c version.c -o version.o\n",
      "g++ -std=gnu++11 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o rlang.so capture.o internal.o rlang-rcc.o rlang.o version.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-rlang/00new/rlang/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (rlang)\n",
      "* installing *source* package ‘cli’ ...\n",
      "** package ‘cli’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ansi.c -o ansi.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c cleancall.c -o cleancall.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c diff.c -o diff.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c errors.c -o errors.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c md5.c -o md5.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c progress-altrep.c -o progress-altrep.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c progress.c -o progress.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c sha256.c -o sha256.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c thread.c -o thread.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c tty.c -o tty.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c utf8.c -o utf8.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c win-utf8.c -o win-utf8.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c winfiles.c -o winfiles.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o cli.so ansi.o cleancall.o diff.o errors.o init.o md5.o progress-altrep.o progress.o sha256.o thread.o tty.o utf8.o win-utf8.o winfiles.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-cli/00new/cli/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (cli)\n",
      "* installing *source* package ‘vctrs’ ...\n",
      "** package ‘vctrs’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c altrep-rle.c -o altrep-rle.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c altrep.c -o altrep.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c arg-counter.c -o arg-counter.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c arg.c -o arg.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c assert.c -o assert.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c bind.c -o bind.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c c-unchop.c -o c-unchop.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c c.c -o c.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c callables.c -o callables.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c cast-bare.c -o cast-bare.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c cast-dispatch.c -o cast-dispatch.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c cast.c -o cast.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c compare.c -o compare.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c complete.c -o complete.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c conditions.c -o conditions.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c dictionary.c -o dictionary.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c dim.c -o dim.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c empty.c -o empty.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c equal.c -o equal.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fields.c -o fields.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fill.c -o fill.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c globals.c -o globals.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c group.c -o group.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c growable.c -o growable.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c hash.c -o hash.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c interval.c -o interval.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c match-joint.c -o match-joint.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c match.c -o match.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c names.c -o names.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c order-collate.c -o order-collate.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c order-groups.c -o order-groups.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c order-sortedness.c -o order-sortedness.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c order-truelength.c -o order-truelength.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c order.c -o order.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c poly-op.c -o poly-op.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c proxy-restore.c -o proxy-restore.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c proxy.c -o proxy.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ptype-common.c -o ptype-common.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ptype.c -o ptype.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ptype2-dispatch.c -o ptype2-dispatch.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ptype2.c -o ptype2.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rank.c -o rank.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rep.c -o rep.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rlang-dev.c -o rlang-dev.o\n",
      "g++ -std=gnu++11 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang     -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rlang-rcc.cpp -o rlang-rcc.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rlang.c -o rlang.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c runs.c -o runs.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c shape.c -o shape.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c size-common.c -o size-common.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c size.c -o size.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c slice-array.c -o slice-array.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c slice-assign-array.c -o slice-assign-array.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c slice-assign.c -o slice-assign.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c slice-chop.c -o slice-chop.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c slice-interleave.c -o slice-interleave.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c slice.c -o slice.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c split.c -o split.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c subscript-loc.c -o subscript-loc.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c subscript.c -o subscript.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c translate.c -o translate.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c type-data-frame.c -o type-data-frame.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c type-date-time.c -o type-date-time.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c type-factor.c -o type-factor.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c type-info.c -o type-info.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c type-integer64.c -o type-integer64.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c type-tibble.c -o type-tibble.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c typeof2-s3.c -o typeof2-s3.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c typeof2.c -o typeof2.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c unspecified.c -o unspecified.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c utils-dispatch.c -o utils-dispatch.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c utils.c -o utils.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c vctrs-core.c -o vctrs-core.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG -I./rlang    -fvisibility=hidden -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c version.c -o version.o\n",
      "g++ -std=gnu++11 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o vctrs.so altrep-rle.o altrep.o arg-counter.o arg.o assert.o bind.o c-unchop.o c.o callables.o cast-bare.o cast-dispatch.o cast.o compare.o complete.o conditions.o dictionary.o dim.o empty.o equal.o fields.o fill.o globals.o group.o growable.o hash.o init.o interval.o match-joint.o match.o names.o order-collate.o order-groups.o order-sortedness.o order-truelength.o order.o poly-op.o proxy-restore.o proxy.o ptype-common.o ptype.o ptype2-dispatch.o ptype2.o rank.o rep.o rlang-dev.o rlang-rcc.o rlang.o runs.o shape.o size-common.o size.o slice-array.o slice-assign-array.o slice-assign.o slice-chop.o slice-interleave.o slice.o split.o subscript-loc.o subscript.o translate.o type-data-frame.o type-date-time.o type-factor.o type-info.o type-integer64.o type-tibble.o typeof2-s3.o typeof2.o unspecified.o utils-dispatch.o utils.o vctrs-core.o version.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-vctrs/00new/vctrs/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (vctrs)\n",
      "* installing *source* package ‘dplyr’ ...\n",
      "** package ‘dplyr’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c chop.cpp -o chop.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c filter.cpp -o filter.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c funs.cpp -o funs.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c group_by.cpp -o group_by.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c group_data.cpp -o group_data.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c imports.cpp -o imports.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.cpp -o init.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c mask.cpp -o mask.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c mutate.cpp -o mutate.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c slice.cpp -o slice.o\n",
      "g++ -std=gnu++14 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c summarise.cpp -o summarise.o\n",
      "g++ -std=gnu++14 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o dplyr.so chop.o filter.o funs.o group_by.o group_data.o imports.o init.o mask.o mutate.o slice.o summarise.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-dplyr/00new/dplyr/libs\n",
      "** R\n",
      "** data\n",
      "*** moving datasets to lazyload DB\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (dplyr)\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: \n",
      "R[write to console]: The downloaded source packages are in\n",
      "\t‘/tmp/RtmpVYkzrU/downloaded_packages’\n",
      "R[write to console]: \n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/data.table_1.14.2.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 5301817 bytes (5.1 MB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 5.1 MB\n",
      "\n",
      "\n",
      "* installing *source* package ‘data.table’ ...\n",
      "** package ‘data.table’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 6.5.0\n",
      "zlib 1.2.11 is available ok\n",
      "R CMD SHLIB supports OpenMP without any extra hint\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c assign.c -o assign.o\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c between.c -o between.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c bmerge.c -o bmerge.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c chmatch.c -o chmatch.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c cj.c -o cj.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c coalesce.c -o coalesce.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c dogroups.c -o dogroups.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fastmean.c -o fastmean.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fcast.c -o fcast.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fifelse.c -o fifelse.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fmelt.c -o fmelt.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c forder.c -o forder.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c frank.c -o frank.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fread.c -o fread.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c freadR.c -o freadR.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c froll.c -o froll.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c frollR.c -o frollR.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c frolladaptive.c -o frolladaptive.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fsort.c -o fsort.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fwrite.c -o fwrite.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c fwriteR.c -o fwriteR.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c gsumm.c -o gsumm.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ijoin.c -o ijoin.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c inrange.c -o inrange.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nafill.c -o nafill.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nqrecreateindices.c -o nqrecreateindices.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c openmp-utils.c -o openmp-utils.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c quickselect.c -o quickselect.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rbindlist.c -o rbindlist.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c reorder.c -o reorder.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c shift.c -o shift.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c snprintf.c -o snprintf.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c subset.c -o subset.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c transpose.c -o transpose.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c types.c -o types.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c uniqlist.c -o uniqlist.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c utils.c -o utils.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c vecseq.c -o vecseq.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG     -fopenmp  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c wrappers.c -o wrappers.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o data.table.so assign.o between.o bmerge.o chmatch.o cj.o coalesce.o dogroups.o fastmean.o fcast.o fifelse.o fmelt.o forder.o frank.o fread.o freadR.o froll.o frollR.o frolladaptive.o fsort.o fwrite.o fwriteR.o gsumm.o ijoin.o init.o inrange.o nafill.o nqrecreateindices.o openmp-utils.o quickselect.o rbindlist.o reorder.o shift.o snprintf.o subset.o transpose.o types.o uniqlist.o utils.o vecseq.o wrappers.o -fopenmp -lz -L/usr/lib/R/lib -lR\n",
      "PKG_CFLAGS = -fopenmp\n",
      "PKG_LIBS = -fopenmp -lz\n",
      "if [ \"data.table.so\" != \"datatable.so\" ]; then mv data.table.so datatable.so; fi\n",
      "if [ \"\" != \"Windows_NT\" ] && [ `uname -s` = 'Darwin' ]; then install_name_tool -id datatable.so datatable.so; fi\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-data.table/00new/data.table/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (data.table)\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: \n",
      "R[write to console]: The downloaded source packages are in\n",
      "\t‘/tmp/RtmpVYkzrU/downloaded_packages’\n",
      "R[write to console]: \n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/arsenal_3.6.3.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 672939 bytes (657 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 657 KB\n",
      "\n",
      "\n",
      "* installing *source* package ‘arsenal’ ...\n",
      "** package ‘arsenal’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** data\n",
      "*** moving datasets to lazyload DB\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (arsenal)\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: \n",
      "R[write to console]: The downloaded source packages are in\n",
      "\t‘/tmp/RtmpVYkzrU/downloaded_packages’\n",
      "R[write to console]: \n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: also installing the dependencies ‘Formula’, ‘viridis’, ‘MatrixModels’, ‘mvtnorm’, ‘TH.data’, ‘sandwich’, ‘checkmate’, ‘Hmisc’, ‘SparseM’, ‘quantreg’, ‘polspline’, ‘multcomp’, ‘htmlTable’, ‘rms’\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/Formula_1.2-4.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 164214 bytes (160 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 160 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/viridis_0.6.2.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 3031731 bytes (2.9 MB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 2.9 MB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/MatrixModels_0.5-0.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 24045 bytes (23 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 23 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/mvtnorm_1.1-3.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 166421 bytes (162 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 162 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/TH.data_1.1-1.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 8612708 bytes (8.2 MB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 8.2 MB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/sandwich_3.0-2.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 1356080 bytes (1.3 MB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 1.3 MB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/checkmate_2.1.0.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 170034 bytes (166 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 166 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/Hmisc_4.7-1.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 792617 bytes (774 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 774 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/SparseM_1.81.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 735100 bytes (717 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 717 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/quantreg_5.94.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 1025290 bytes (1001 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 1001 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/polspline_1.1.20.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 262851 bytes (256 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 256 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/multcomp_1.4-20.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 615868 bytes (601 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 601 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/htmlTable_2.4.1.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 439117 bytes (428 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 428 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/rms_6.3-0.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 586546 bytes (572 KB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 572 KB\n",
      "\n",
      "\n",
      "R[write to console]: trying URL 'https://cloud.r-project.org/src/contrib/haplo.stats_1.8.9.tar.gz'\n",
      "\n",
      "R[write to console]: Content type 'application/x-gzip'\n",
      "R[write to console]:  length 2755760 bytes (2.6 MB)\n",
      "\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: =\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: downloaded 2.6 MB\n",
      "\n",
      "\n",
      "* installing *source* package ‘Formula’ ...\n",
      "** package ‘Formula’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (Formula)\n",
      "* installing *source* package ‘viridis’ ...\n",
      "** package ‘viridis’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** data\n",
      "*** moving datasets to lazyload DB\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (viridis)\n",
      "* installing *source* package ‘MatrixModels’ ...\n",
      "** package ‘MatrixModels’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (MatrixModels)\n",
      "* installing *source* package ‘mvtnorm’ ...\n",
      "** package ‘mvtnorm’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c C_FORTRAN_interface.c -o C_FORTRAN_interface.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c miwa.c -o miwa.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c mvt.f -o mvt.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c mvtnorm-init.c -o mvtnorm-init.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c tvpack.f -o tvpack.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o mvtnorm.so C_FORTRAN_interface.o miwa.o mvt.o mvtnorm-init.o tvpack.o -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-mvtnorm/00new/mvtnorm/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (mvtnorm)\n",
      "* installing *source* package ‘TH.data’ ...\n",
      "** package ‘TH.data’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** data\n",
      "*** moving datasets to lazyload DB\n",
      "** demo\n",
      "** inst\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (TH.data)\n",
      "* installing *source* package ‘sandwich’ ...\n",
      "** package ‘sandwich’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** data\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "*** copying figures\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (sandwich)\n",
      "* installing *source* package ‘checkmate’ ...\n",
      "** package ‘checkmate’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c all_missing.c -o all_missing.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c any_infinite.c -o any_infinite.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c any_missing.c -o any_missing.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c any_nan.c -o any_nan.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c checks.c -o checks.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c find_nchar.c -o find_nchar.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c guess_type.c -o guess_type.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c helper.c -o helper.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c integerish.c -o integerish.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c is_sorted.c -o is_sorted.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c qassert.c -o qassert.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c which_first.c -o which_first.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o checkmate.so all_missing.o any_infinite.o any_missing.o any_nan.o checks.o find_nchar.o guess_type.o helper.o init.o integerish.o is_sorted.o qassert.o which_first.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-checkmate/00new/checkmate/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (checkmate)\n",
      "* installing *source* package ‘SparseM’ ...\n",
      "** package ‘SparseM’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c bckslv.f -o bckslv.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c chol.f -o chol.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c chol2csr.f -o chol2csr.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c cholesky.f -o cholesky.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c csr.f -o csr.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c extract.f -o extract.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c sparskit.f -o sparskit.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c subscr.f -o subscr.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o SparseM.so bckslv.o chol.o chol2csr.o cholesky.o csr.o extract.o init.o sparskit.o subscr.o -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-SparseM/00new/SparseM/libs\n",
      "** R\n",
      "** data\n",
      "** demo\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating a generic function for ‘diag’ from package ‘base’ in package ‘SparseM’\n",
      "Creating a generic function for ‘diag<-’ from package ‘base’ in package ‘SparseM’\n",
      "Creating a generic function for ‘norm’ from package ‘base’ in package ‘SparseM’\n",
      "Creating a new generic function for ‘backsolve’ in package ‘SparseM’\n",
      "Creating a generic function for ‘forwardsolve’ from package ‘base’ in package ‘SparseM’\n",
      "Creating a generic function for ‘model.response’ from package ‘stats’ in package ‘SparseM’\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (SparseM)\n",
      "* installing *source* package ‘polspline’ ...\n",
      "** package ‘polspline’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c allpack.f -o allpack.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c hareall.c -o hareall.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c heftall.c -o heftall.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c lsdall.c -o lsdall.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c lspecall.c -o lspecall.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nlsd.c -o nlsd.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c polyall.c -o polyall.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c polymars.c -o polymars.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c registerDynamicSymbol.c -o registerDynamicSymbol.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o polspline.so allpack.o hareall.o heftall.o lsdall.o lspecall.o nlsd.o polyall.o polymars.o registerDynamicSymbol.o -lblas -lgfortran -lm -lquadmath -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-polspline/00new/polspline/libs\n",
      "** R\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (polspline)\n",
      "* installing *source* package ‘quantreg’ ...\n",
      "** package ‘quantreg’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c akj.f -o akj.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c boot.f -o boot.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c bound.f -o bound.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c boundc.f -o boundc.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c brute.f -o brute.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c chlfct.f -o chlfct.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c cholesky.f -o cholesky.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c combos.f -o combos.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c crqf.f -o crqf.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c crqfnb.f -o crqfnb.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c dsel05.f -o dsel05.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c etime.f -o etime.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c extract.f -o extract.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c frand.c -o frand.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c grexp.f -o grexp.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c idmin.f -o idmin.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c iswap.f -o iswap.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c kuantile.f -o kuantile.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c kuantiles.f -o kuantiles.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c linpack.f -o linpack.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c mcmb.c -o mcmb.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c penalty.f -o penalty.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c pfnb.f -o pfnb.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c powell.f -o powell.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c profnb.f -o profnb.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c pwxy.f -o pwxy.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c qfnb.f -o qfnb.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c qselect.f -o qselect.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c quantreg_init.c -o quantreg_init.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rls.f -o rls.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rq0.f -o rq0.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rq1.f -o rq1.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rqbr.f -o rqbr.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rqfn.f -o rqfn.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rqfnb.f -o rqfnb.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rqfnc.f -o rqfnc.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rqs.f -o rqs.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c sakj.f -o sakj.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c sparskit2.f -o sparskit2.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c srqfn.f -o srqfn.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c srqfnc.f -o srqfnc.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c srtpai.f -o srtpai.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o quantreg.so akj.o boot.o bound.o boundc.o brute.o chlfct.o cholesky.o combos.o crqf.o crqfnb.o dsel05.o etime.o extract.o frand.o grexp.o idmin.o iswap.o kuantile.o kuantiles.o linpack.o mcmb.o penalty.o pfnb.o powell.o profnb.o pwxy.o qfnb.o qselect.o quantreg_init.o rls.o rq0.o rq1.o rqbr.o rqfn.o rqfnb.o rqfnc.o rqs.o sakj.o sparskit2.o srqfn.o srqfnc.o srtpai.o -llapack -lblas -lgfortran -lm -lquadmath -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-quantreg/00new/quantreg/libs\n",
      "** R\n",
      "** data\n",
      "** demo\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (quantreg)\n",
      "* installing *source* package ‘multcomp’ ...\n",
      "** package ‘multcomp’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** data\n",
      "*** moving datasets to lazyload DB\n",
      "** demo\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (multcomp)\n",
      "* installing *source* package ‘htmlTable’ ...\n",
      "** package ‘htmlTable’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** R\n",
      "** data\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (htmlTable)\n",
      "* installing *source* package ‘Hmisc’ ...\n",
      "** package ‘Hmisc’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c Hmisc.c -o Hmisc.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c cidxcn.f -o cidxcn.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c cidxcp.f -o cidxcp.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c hoeffd.f -o hoeffd.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c jacklins.f -o jacklins.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c largrec.f -o largrec.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c mChoice.c -o mChoice.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c maxempr.f -o maxempr.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c nstr.c -o nstr.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c ranksort.c -o ranksort.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c rcorr.f -o rcorr.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c string_box.c -o string_box.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c wclosest.f -o wclosest.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o Hmisc.so Hmisc.o cidxcn.o cidxcp.o hoeffd.o init.o jacklins.o largrec.o mChoice.o maxempr.o nstr.o ranksort.o rcorr.o string_box.o wclosest.o -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-Hmisc/00new/Hmisc/libs\n",
      "** R\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (Hmisc)\n",
      "* installing *source* package ‘rms’ ...\n",
      "** package ‘rms’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c lrmfit.f -o lrmfit.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c mlmats.f -o mlmats.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c ormuv.f -o ormuv.o\n",
      "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong  -c robcovf.f -o robcovf.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o rms.so init.o lrmfit.o mlmats.o ormuv.o robcovf.o -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-rms/00new/rms/libs\n",
      "** R\n",
      "** demo\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (rms)\n",
      "* installing *source* package ‘haplo.stats’ ...\n",
      "** package ‘haplo.stats’ successfully unpacked and MD5 sums checked\n",
      "** using staged installation\n",
      "** libs\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c groupsum.c -o groupsum.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c haplo.stats_init.c -o haplo.stats_init.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c haplo_em_pin.c -o haplo_em_pin.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c louis_info.c -o louis_info.o\n",
      "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-5XUBcI/r-base-4.1.1=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c seqhap.c -o seqhap.o\n",
      "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o haplo.stats.so groupsum.o haplo.stats_init.o haplo_em_pin.o louis_info.o seqhap.o -L/usr/lib/R/lib -lR\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing to /home/jupyter/R_packages/00LOCK-haplo.stats/00new/haplo.stats/libs\n",
      "** R\n",
      "** data\n",
      "** inst\n",
      "** byte-compile and prepare package for lazy loading\n",
      "** help\n",
      "*** installing help indices\n",
      "** building package indices\n",
      "** installing vignettes\n",
      "** testing if installed package can be loaded from temporary location\n",
      "** checking absolute paths in shared objects and dynamic libraries\n",
      "** testing if installed package can be loaded from final location\n",
      "** testing if installed package keeps a record of temporary installation path\n",
      "* DONE (haplo.stats)\n",
      "R[write to console]: \n",
      "\n",
      "R[write to console]: \n",
      "R[write to console]: The downloaded source packages are in\n",
      "\t‘/tmp/RtmpVYkzrU/downloaded_packages’\n",
      "R[write to console]: \n",
      "R[write to console]: \n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "pack <- \"/home/jupyter/R_packages\"\n",
    "\n",
    "install.packages(\"dplyr\", lib = pack)\n",
    "install.packages(\"data.table\", lib = pack)\n",
    "install.packages(\"arsenal\", lib = pack)\n",
    "install.packages(\"haplo.stats\", lib = pack)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Load packages:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "pack <- \"/home/jupyter/R_packages\"\n",
    "suppressPackageStartupMessages(library(dplyr, lib.loc = pack))\n",
    "suppressPackageStartupMessages(library(data.table, lib.loc = pack))\n",
    "suppressPackageStartupMessages(library(arsenal, lib.loc = pack))\n",
    "suppressPackageStartupMessages(library(haplo.stats, lib.loc = pack))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* Add SNP name to the PED file:\n",
    "- Since the .ped file contain 6+2V fields (where V is the number of variants) but where the variant names are in the .map file, we need to create a list with the rsIDs/SNP names for all SNPs, with 2 alleles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "PED <- fread(\"/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.ped\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "FILTERED.SNCA_1 <- read.table(\"/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.map\", header=FALSE)\n",
    "FILTERED.SNCA_2 <- read.table(\"/home/jupyter/1000G_TEST/SNCA_hap/FILTERED.test_formatted.SNCA.map\", header=FALSE)\n",
    "FILTERED.SNCA_1$V2 <- paste0(FILTERED.SNCA_1$V2, sep=\"_\", \"1\")\n",
    "FILTERED.SNCA_2$V2 <- paste0(FILTERED.SNCA_2$V2, sep=\"_\", \"2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#Set column names\n",
    "colnames(FILTERED.SNCA_1) <- c(\"CHR\", \"SNP\", \"CM\", \"POS\")\n",
    "colnames(FILTERED.SNCA_2) <- c(\"CHR\", \"SNP\", \"CM\", \"POS\")\n",
    "FILTERED.SNCA_2alleles <- rbind(FILTERED.SNCA_1, FILTERED.SNCA_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#Sort file and create geno matrix:\n",
    "#The variants should appear in order of genomic postition in the ped file, therefore, we sort the alleles based on position\n",
    "SNCA_pos <- FILTERED.SNCA_2alleles[order(FILTERED.SNCA_2alleles$POS),]\n",
    "SNCA_alleles <- SNCA_pos[,c(\"SNP\")]\n",
    "write.table(SNCA_alleles, file=\"/home/jupyter/1000G_TEST/SNCA_hap/SNCA_alleles.txt\", quote = F, sep = \"\\t\", row.names = F, col.names = F)\n",
    "\n",
    "#Add allele names to the PED file\n",
    "colnames(PED) <- c(\"FID\", \"IID\", \"PAT\",\"MAT\", \"SEX\", \"PHENO\", SNCA_alleles)\n",
    "write.table(PED, file=\"/home/jupyter/1000G_TEST/SNCA_hap/Geno_matrix_SNCA.tab\", quote = F, row.names = F, sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#Import covariate file and set which variables you would like to adjust for\n",
    "SampleInfo_Adjustment <- read.delim(\"/home/jupyter/1000G_TEST/covariateFile_Haplo.txt\")\n",
    "adj <- data.frame(SampleInfo_Adjustment[,c(\"SEX\",\"AGE\",\"PC1\", \"PC2\", \"PC3\", \"PC4\", \"PC5\")])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Create haplotype block files (N=4 blocks)**\n",
    "\n",
    "Information on the different variants in each haplotype block was retrieved from the .blocks file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* H1 (Haplotype block 1)\n",
    "- SNPs in this haplotype block (N=5): 4:90645674, 4:90653134, 4:90663542, 4:90663670, 4:90668019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "Geno_matrix_SNCA <- read.delim(\"/home/jupyter/1000G_TEST/SNCA_hap/Geno_matrix_SNCA.tab\", check.names = FALSE)\n",
    "#Colnames:\n",
    "SNCA_alleles <- read.table(\"/home/jupyter/1000G_TEST/SNCA_hap/SNCA_alleles.txt\", quote=\"\\\"\", comment.char=\"\")\n",
    "H1_a <- SNCA_alleles[grepl(\"4:90645674|4:90653134|4:90663542|4:90663670|4:90668019\", SNCA_alleles$V1), ]\n",
    "H1_SNCA <- Geno_matrix_SNCA[,c(\"FID\", \"IID\", \"PAT\", \"MAT\", \"SEX\", \"PHENO\", H1_a)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#Extract only genotype data:\n",
    "SNCA_alleles <- readLines(\"/home/jupyter/1000G_TEST/SNCA_hap/Geno_matrix_SNCA.tab\")\n",
    "geno <- data.frame(H1_SNCA[,c(7:16)], check.names = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#Set variables for running the association analyses in haplo.stats\n",
    "\n",
    "#Label the SNPs:\n",
    "label <- c(\"4_90645674\", \"4_90653134\", \"4_90663542\", \"4_90663670\", \"4_90668019\")\n",
    "#Set binary pheno (0=control, 1=patient):\n",
    "H1_SNCA$PHENO_01 <- H1_SNCA$PHENO-1\n",
    "y.bin <- 1*(H1_SNCA$PHENO_01==\"1\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Run association analyses (both non-adjusted and adjusted) for each haplotype**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 2.6, df = 4, p-val = 0.64\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90645674 4_90653134 4_90663542 4_90663670 4_90668019 Hap-Score p-val\n",
      "1           C          C          C          T          A     -0.77  0.44\n",
      "2           C          C          T          T          A     -0.15  0.88\n",
      "8           T          C          T          G          G      0.15  0.88\n",
      "7           T          C          T          G          A      0.91  0.36\n",
      "4           C          T          T          T          A      1.22  0.22\n",
      "3           C          T          C          T          A        NA    NA\n",
      "5           T          C          C          G          A        NA    NA\n",
      "6           T          C          C          G          G        NA    NA\n",
      "9           T          T          T          G          G        NA    NA\n",
      "NA       <NA>       <NA>       <NA>       <NA>       <NA>        NA    NA\n",
      "   pool.hf control.hf case.hf glm.eff OR.lower  OR OR.upper\n",
      "1  5.3e-01    5.4e-01 5.1e-01    Base       NA 1.0       NA\n",
      "2  3.3e-01    3.3e-01 3.3e-01     Eff     0.77 1.0      1.4\n",
      "8  5.6e-02    5.4e-02 5.7e-02     Eff     0.58 1.1      2.0\n",
      "7  1.1e-02    7.4e-03 1.5e-02     Eff     0.52 1.9      7.0\n",
      "4  7.3e-02    6.2e-02 8.4e-02     Eff     0.83 1.4      2.5\n",
      "3  8.2e-10    5.2e-09      NA    <NA>       NA  NA       NA\n",
      "5  5.1e-12    2.5e-10      NA    <NA>       NA  NA       NA\n",
      "6  4.2e-11         NA      NA    <NA>       NA  NA       NA\n",
      "9  1.2e-09         NA 1.4e-08    <NA>       NA  NA       NA\n",
      "NA      NA         NA      NA    <NA>       NA  NA       NA\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "#Non-adjusted:\n",
    "H1 <- haplo.cc(y=y.bin, geno=geno, locus.label= label, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H1, nlines=10, digits=2)\n",
    "#Sort the output on p-value:\n",
    "H1_cc.df <- H1$cc.df\n",
    "H1_cc.df_sort <- H1_cc.df[order(H1_cc.df$`p-val`),]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 0.4, df = 4, p-val = 0.98\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90645674 4_90653134 4_90663542 4_90663670 4_90668019 Hap-Score p-val\n",
      "4           C          T          T          T          A     -0.34  0.73\n",
      "2           C          C          T          T          A     -0.16  0.87\n",
      "8           T          C          T          G          G      0.10  0.92\n",
      "1           C          C          C          T          A      0.23  0.82\n",
      "7           T          C          T          G          A      0.50  0.62\n",
      "3           C          T          C          T          A        NA    NA\n",
      "5           T          C          C          G          A        NA    NA\n",
      "6           T          C          C          G          G        NA    NA\n",
      "9           T          T          T          G          G        NA    NA\n",
      "NA       <NA>       <NA>       <NA>       <NA>       <NA>        NA    NA\n",
      "   pool.hf control.hf case.hf glm.eff OR.lower   OR OR.upper\n",
      "4  7.3e-02    6.2e-02 8.4e-02     Eff     0.32 0.84      2.2\n",
      "2  3.3e-01    3.3e-01 3.3e-01     Eff     0.55 0.95      1.6\n",
      "8  5.6e-02    5.4e-02 5.7e-02     Eff     0.35 1.02      3.0\n",
      "1  5.3e-01    5.4e-01 5.1e-01    Base       NA 1.00       NA\n",
      "7  1.1e-02    7.4e-03 1.5e-02     Eff     0.03 3.21    347.2\n",
      "3  8.2e-10    5.2e-09      NA    <NA>       NA   NA       NA\n",
      "5  5.1e-12    2.5e-10      NA    <NA>       NA   NA       NA\n",
      "6  4.2e-11         NA      NA    <NA>       NA   NA       NA\n",
      "9  1.2e-09         NA 1.4e-08    <NA>       NA   NA       NA\n",
      "NA      NA         NA      NA    <NA>       NA   NA       NA\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "#Adjusted for age, sex, PC1-5:\n",
    "H1_adj <- haplo.cc(y=y.bin, geno=geno, locus.label= label, x.adj=adj, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H1_adj, nlines=10, digits=2)\n",
    "H1_adj_cc.df <- H1_adj$cc.df\n",
    "H1_adj_cc.df_sort <- H1_adj_cc.df[order(H1_adj_cc.df$`p-val`),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here we tested the association between PD and the identified haplotypes in haplotype block 1, first non-adjusted and then adjusted for age, sex and PC1-5. Each line represent a haplotype in the block and the base for each SNP that is in it. A MAF>1% was used, therefore, some haplotypes have not been includd in the analyses (rows containing NA's), control.hf = haplotype frequency for the control group, case.hf = haplotype frequency for the patient group. \n",
    "\n",
    "We will continue with the rest of the haplotype blocks below:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* H2 (Haplotype block 2)\n",
    "- SNPs in this haplotype block (N=7): 4:90668614 4:90672457 4:90678541 4:90681236 4:90684122 4:90690329 4:90709741"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 3.8, df = 5, p-val = 0.58\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90668614 4_90672457 4_90678541 4_90681236 4_90684122 4_90690329 4_90709741\n",
      "4           C          A          A          C          A          C          G\n",
      "12          T          T          G          A          A          C          A\n",
      "2           C          A          A          A          A          C          G\n",
      "10          T          A          G          A          G          T          A\n",
      "7           T          A          G          A          A          C          G\n",
      "1           C          A          A          A          A          C          A\n",
      "3           C          A          A          A          A          T          G\n",
      "5           C          A          A          C          A          T          A\n",
      "6           C          T          A          C          G          T          A\n",
      "8           T          A          G          A          A          T          A\n",
      "   Hap-Score p-val pool.hf control.hf case.hf glm.eff OR.lower   OR OR.upper\n",
      "4     -1.476  0.14 5.4e-02     0.0651 4.2e-02     Eff     0.35 0.65      1.2\n",
      "12    -0.768  0.44 5.3e-02     0.0594 4.7e-02     Eff     0.44 0.83      1.6\n",
      "2     -0.168  0.87 4.7e-01     0.4745 4.7e-01    Base       NA 1.00       NA\n",
      "10    -0.042  0.97 5.4e-02     0.0545 5.2e-02     Eff     0.50 0.95      1.8\n",
      "7      1.039  0.30 3.5e-01     0.3317 3.7e-01     Eff     0.81 1.11      1.5\n",
      "1         NA    NA 1.2e-03         NA 2.5e-03       R     0.52 1.53      4.5\n",
      "3         NA    NA      NA         NA 2.6e-07    <NA>       NA   NA       NA\n",
      "5         NA    NA 9.7e-09         NA      NA    <NA>       NA   NA       NA\n",
      "6         NA    NA 1.2e-03         NA 2.5e-03       R     0.52 1.53      4.5\n",
      "8         NA    NA 7.0e-03     0.0057 7.4e-03       R     0.52 1.53      4.5\n",
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 3.7, df = 5, p-val = 0.6\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90668614 4_90672457 4_90678541 4_90681236 4_90684122 4_90690329 4_90709741\n",
      "4           C          A          A          C          A          C          G\n",
      "12          T          T          G          A          A          C          A\n",
      "10          T          A          G          A          G          T          A\n",
      "7           T          A          G          A          A          C          G\n",
      "2           C          A          A          A          A          C          G\n",
      "1           C          A          A          A          A          C          A\n",
      "3           C          A          A          A          A          T          G\n",
      "5           C          A          A          C          A          T          A\n",
      "6           C          T          A          C          G          T          A\n",
      "8           T          A          G          A          A          T          A\n",
      "   Hap-Score p-val pool.hf control.hf case.hf glm.eff OR.lower   OR OR.upper\n",
      "4     -1.326  0.18 5.4e-02     0.0651 4.2e-02     Eff     0.13 0.44      1.5\n",
      "12    -1.074  0.28 5.3e-02     0.0594 4.7e-02     Eff     0.14 0.49      1.7\n",
      "10     0.039  0.97 5.4e-02     0.0545 5.2e-02     Eff     0.28 0.84      2.6\n",
      "7      0.122  0.90 3.5e-01     0.3317 3.7e-01     Eff     0.54 0.93      1.6\n",
      "2      0.736  0.46 4.7e-01     0.4745 4.7e-01    Base       NA 1.00       NA\n",
      "1         NA    NA 1.2e-03         NA 2.5e-03       R     0.18 2.86     44.6\n",
      "3         NA    NA      NA         NA 2.6e-07    <NA>       NA   NA       NA\n",
      "5         NA    NA 9.7e-09         NA      NA    <NA>       NA   NA       NA\n",
      "6         NA    NA 1.2e-03         NA 2.5e-03       R     0.18 2.86     44.6\n",
      "8         NA    NA 7.0e-03     0.0057 7.4e-03       R     0.18 2.86     44.6\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "#Run the same analyses as above for haplotype block 2:\n",
    "\n",
    "#Colnames:\n",
    "SNCA_alleles <- read.table(\"/home/jupyter/1000G_TEST/SNCA_hap/SNCA_alleles.txt\", quote=\"\\\"\", comment.char=\"\")\n",
    "H2_a <- SNCA_alleles[grepl(\"4:90668614|4:90672457|4:90678541|4:90681236|4:90684122|4:90690329|4:90709741\", SNCA_alleles$V1), ]\n",
    "H2_SNCA <- Geno_matrix_SNCA[,c(\"FID\", \"IID\", \"PAT\", \"MAT\", \"SEX\", \"PHENO\", H2_a)]\n",
    "\n",
    "#Extract only genotype data:\n",
    "geno <- data.frame(H2_SNCA[,c(7:20)], check.names = FALSE)\n",
    "\n",
    "#Label the SNPs:\n",
    "label <- c(\"4_90668614\",\"4_90672457\",\"4_90678541\",\"4_90681236\",\"4_90684122\",\"4_90690329\",\"4_90709741\")\n",
    "\n",
    "#Set binary pheno (0 ctrl, 1 pat):\n",
    "H2_SNCA$PHENO_01 <- H2_SNCA$PHENO-1\n",
    "y.bin <- 1*(H2_SNCA$PHENO_01==\"1\")\n",
    "\n",
    "#Non-adjusted:\n",
    "H2 <- haplo.cc(y=y.bin, geno=geno, locus.label= label, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H2, nlines=10, digits=2)\n",
    "\n",
    "#Sort the output on p-value:\n",
    "H2_cc.df <- H2$cc.df\n",
    "H2_cc.df_sort <- H2_cc.df[order(H2_cc.df$`p-val`),]\n",
    "\n",
    "#Adjusted for age, sex, PC1-5:\n",
    "H2_adj <- haplo.cc(y=y.bin, geno=geno, locus.label= label, x.adj=adj, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H2_adj, nlines=10, digits=2)\n",
    "H2_adj_cc.df <- H2_adj$cc.df\n",
    "H2_adj_cc.df_sort <- H2_adj_cc.df[order(H2_adj_cc.df$`p-val`),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* H3 (Haplotype block 3)\n",
    "- SNPs in this haplotype block (N=7): 4:90716852 4:90718995 4:90722145 4:90734535 4:90736006 4:90736113 4:90739539"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 3.9, df = 5, p-val = 0.56\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90716852 4_90718995 4_90722145 4_90734535 4_90736006 4_90736113 4_90739539\n",
      "8           T          C          A          T          A          T          C\n",
      "2           C          C          C          T          A          C          C\n",
      "6           T          C          A          G          C          C          A\n",
      "9           T          C          A          T          C          C          A\n",
      "10          T          C          A          T          C          C          C\n",
      "13          T          T          A          T          C          C          A\n",
      "1           C          C          C          T          A          C          A\n",
      "3           C          T          C          T          A          C          A\n",
      "4           C          T          C          T          A          C          C\n",
      "5           T          C          A          G          A          T          A\n",
      "   Hap-Score p-val pool.hf control.hf case.hf glm.eff OR.lower   OR OR.upper\n",
      "8      -1.65 0.098 2.1e-01    2.3e-01 1.8e-01     Eff     0.49 0.74      1.1\n",
      "2      -0.70 0.482 6.3e-02    6.9e-02 5.7e-02     Eff     0.44 0.79      1.4\n",
      "6       0.30 0.766 5.4e-02    5.2e-02 5.7e-02     Eff     0.54 1.00      1.9\n",
      "9       0.33 0.738 2.4e-01    2.3e-01 2.4e-01     Eff     0.64 0.95      1.4\n",
      "10      0.61 0.539 3.0e-01    2.9e-01 3.1e-01    Base       NA 1.00       NA\n",
      "13      1.04 0.299 1.4e-01    1.3e-01 1.6e-01     Eff     0.72 1.13      1.8\n",
      "1         NA    NA 3.5e-11         NA 8.4e-10    <NA>       NA   NA       NA\n",
      "3         NA    NA 1.9e-10    1.7e-09      NA    <NA>       NA   NA       NA\n",
      "4         NA    NA      NA    4.9e-11      NA    <NA>       NA   NA       NA\n",
      "5         NA    NA      NA    5.9e-10      NA    <NA>       NA   NA       NA\n",
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 4.2, df = 5, p-val = 0.52\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90716852 4_90718995 4_90722145 4_90734535 4_90736006 4_90736113 4_90739539\n",
      "8           T          C          A          T          A          T          C\n",
      "6           T          C          A          G          C          C          A\n",
      "10          T          C          A          T          C          C          C\n",
      "2           C          C          C          T          A          C          C\n",
      "9           T          C          A          T          C          C          A\n",
      "13          T          T          A          T          C          C          A\n",
      "1           C          C          C          T          A          C          A\n",
      "3           C          T          C          T          A          C          A\n",
      "4           C          T          C          T          A          C          C\n",
      "5           T          C          A          G          A          T          A\n",
      "   Hap-Score p-val pool.hf control.hf case.hf glm.eff OR.lower   OR OR.upper\n",
      "8      -1.61  0.11 2.1e-01    2.3e-01 1.8e-01     Eff     0.30 0.68      1.5\n",
      "6      -0.43  0.67 5.4e-02    5.2e-02 5.7e-02     Eff     0.27 0.84      2.6\n",
      "10     -0.33  0.74 3.0e-01    2.9e-01 3.1e-01    Base       NA 1.00       NA\n",
      "2      -0.28  0.78 6.3e-02    6.9e-02 5.7e-02     Eff     0.28 0.93      3.1\n",
      "9       1.07  0.29 2.4e-01    2.3e-01 2.4e-01     Eff     0.65 1.25      2.4\n",
      "13      1.27  0.21 1.4e-01    1.3e-01 1.6e-01     Eff     0.66 1.51      3.5\n",
      "1         NA    NA 3.5e-11         NA 8.4e-10    <NA>       NA   NA       NA\n",
      "3         NA    NA 1.9e-10    1.7e-09      NA    <NA>       NA   NA       NA\n",
      "4         NA    NA      NA    4.9e-11      NA    <NA>       NA   NA       NA\n",
      "5         NA    NA      NA    5.9e-10      NA    <NA>       NA   NA       NA\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "#Colnames:\n",
    "SNCA_alleles <- read.table(\"/home/jupyter/1000G_TEST/SNCA_hap/SNCA_alleles.txt\", quote=\"\\\"\", comment.char=\"\")\n",
    "H3_a <- SNCA_alleles[grepl(\"4:90716852|4:90718995|4:90722145|4:90734535|4:90736006|4:90736113|4:90739539\", SNCA_alleles$V1), ]\n",
    "H3_SNCA <- Geno_matrix_SNCA[,c(\"FID\", \"IID\", \"PAT\", \"MAT\", \"SEX\", \"PHENO\", H3_a)]\n",
    "\n",
    "#Extract only genotype data:\n",
    "geno <- data.frame(H3_SNCA[,c(7:20)], check.names = FALSE)\n",
    "\n",
    "#Label the SNPs:\n",
    "label <- c(\"4_90716852\",\"4_90718995\",\"4_90722145\",\"4_90734535\",\"4_90736006\",\"4_90736113\",\"4_90739539\")\n",
    "\n",
    "#Set binary pheno (0 ctrl, 1 pat):\n",
    "H3_SNCA$PHENO_01 <- H3_SNCA$PHENO-1\n",
    "y.bin <- 1*(H3_SNCA$PHENO_01==\"1\")\n",
    "\n",
    "#Non-adjusted:\n",
    "H3 <- haplo.cc(y=y.bin, geno=geno, locus.label= label, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H3, nlines=10, digits=2)\n",
    "\n",
    "#Sort the output on p-value:\n",
    "H3_cc.df <- H3$cc.df\n",
    "H3_cc.df_sort <- H3_cc.df[order(H3_cc.df$`p-val`),]\n",
    "\n",
    "#Adjusted for age, sex, PC1-5:\n",
    "H3_adj <- haplo.cc(y=y.bin, geno=geno, locus.label= label, x.adj=adj, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H3_adj, nlines=10, digits=2)\n",
    "H3_adj_cc.df <- H3_adj$cc.df\n",
    "H3_adj_cc.df_sort <- H3_adj_cc.df[order(H3_adj_cc.df$`p-val`),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* H4 (Haplotype block 4)\n",
    "- SNPs in this haplotype block (N=8): 4:90753339 4:90754292 4:90755939 4:90757272 4:90757735 4:90757840 4:90757947 4:90758945"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 2.9, df = 7, p-val = 0.89\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90753339 4_90754292 4_90755939 4_90757272 4_90757735 4_90757840 4_90757947\n",
      "7           A          T          A          G          C          T          T\n",
      "9           A          T          G          G          A          C          T\n",
      "6           A          T          A          G          C          C          T\n",
      "4           A          C          A          G          C          T          T\n",
      "13          T          T          G          G          A          C          A\n",
      "8           A          T          G          G          A          C          A\n",
      "2           A          C          A          A          C          C          T\n",
      "1           A          C          A          A          C          C          T\n",
      "3           A          C          A          A          C          T          T\n",
      "5           A          T          A          A          C          T          T\n",
      "   4_90758945 Hap-Score p-val pool.hf control.hf case.hf glm.eff OR.lower   OR\n",
      "7           A     -1.37  0.17  0.1889     0.2079  0.1701     Eff     0.54 0.79\n",
      "9           A     -0.82  0.41  0.0112     0.0172  0.0080     Eff     0.13 0.54\n",
      "6           A     -0.31  0.76  0.0136     0.0149  0.0124     Eff     0.23 0.76\n",
      "4           A     -0.17  0.86  0.0166     0.0173  0.0155     Eff     0.28 0.87\n",
      "13          A      0.25  0.80  0.0583     0.0593  0.0599     Eff     0.54 1.01\n",
      "8           A      0.64  0.52  0.2202     0.2080  0.2297     Eff     0.74 1.05\n",
      "2           G      0.69  0.49  0.4802     0.4678  0.4926    Base       NA 1.00\n",
      "1           A        NA    NA  0.0012     0.0025      NA       R     0.24 1.09\n",
      "3           A        NA    NA  0.0025     0.0025  0.0025       R     0.24 1.09\n",
      "5           A        NA    NA  0.0012     0.0025      NA       R     0.24 1.09\n",
      "   OR.upper\n",
      "7       1.1\n",
      "9       2.2\n",
      "6       2.5\n",
      "4       2.7\n",
      "13      1.9\n",
      "8       1.5\n",
      "2        NA\n",
      "1       4.9\n",
      "3       4.9\n",
      "5       4.9\n",
      "-------------------------------------------------------------------------------- \n",
      "                            Global Score Statistics                              \n",
      "-------------------------------------------------------------------------------- \n",
      "global-stat = 7.8, df = 7, p-val = 0.35\n",
      "\n",
      "-------------------------------------------------------------------------------- \n",
      "                         Counts for Cases and Controls                           \n",
      "-------------------------------------------------------------------------------- \n",
      "control    case \n",
      "    202     202 \n",
      "\n",
      "\n",
      "   4_90753339 4_90754292 4_90755939 4_90757272 4_90757735 4_90757840 4_90757947\n",
      "7           A          T          A          G          C          T          T\n",
      "8           A          T          G          G          A          C          A\n",
      "9           A          T          G          G          A          C          T\n",
      "6           A          T          A          G          C          C          T\n",
      "4           A          C          A          G          C          T          T\n",
      "2           A          C          A          A          C          C          T\n",
      "13          T          T          G          G          A          C          A\n",
      "1           A          C          A          A          C          C          T\n",
      "3           A          C          A          A          C          T          T\n",
      "5           A          T          A          A          C          T          T\n",
      "   4_90758945 Hap-Score p-val pool.hf control.hf case.hf glm.eff OR.lower   OR\n",
      "7           A     -1.16  0.25  0.1889     0.2079  0.1701     Eff    0.323 0.67\n",
      "8           A     -0.93  0.35  0.2202     0.2080  0.2297     Eff    0.399 0.76\n",
      "9           A     -0.90  0.37  0.0112     0.0172  0.0080     Eff    0.014 0.22\n",
      "6           A     -0.23  0.82  0.0136     0.0149  0.0124     Eff    0.039 0.68\n",
      "4           A      0.38  0.71  0.0166     0.0173  0.0155     Eff    0.207 1.15\n",
      "2           G      0.92  0.36  0.4802     0.4678  0.4926    Base       NA 1.00\n",
      "13          A      0.95  0.34  0.0583     0.0593  0.0599     Eff    0.366 1.37\n",
      "1           A        NA    NA  0.0012     0.0025      NA       R    0.675 7.91\n",
      "3           A        NA    NA  0.0025     0.0025  0.0025       R    0.675 7.91\n",
      "5           A        NA    NA  0.0012     0.0025      NA       R    0.675 7.91\n",
      "   OR.upper\n",
      "7       1.4\n",
      "8       1.4\n",
      "9       3.4\n",
      "6      11.6\n",
      "4       6.4\n",
      "2        NA\n",
      "13      5.1\n",
      "1      92.6\n",
      "3      92.6\n",
      "5      92.6\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "#Colnames:\n",
    "SNCA_alleles <- read.table(\"/home/jupyter/1000G_TEST/SNCA_hap/SNCA_alleles.txt\", quote=\"\\\"\", comment.char=\"\")\n",
    "H4_a <- SNCA_alleles[grepl(\"4:90753339|4:90754292|4:90755939|4:90757272|4:90757735|4:90757840|4:90757947|4:90758945\", SNCA_alleles$V1), ]\n",
    "H4_SNCA <- Geno_matrix_SNCA[,c(\"FID\", \"IID\", \"PAT\", \"MAT\", \"SEX\", \"PHENO\", H4_a)]\n",
    "\n",
    "#Extract only genotype data:\n",
    "geno <- data.frame(H4_SNCA[,c(7:22)], check.names = FALSE)\n",
    "\n",
    "#Label the SNPs:\n",
    "label <- c(\"4_90753339\",\"4_90754292\",\"4_90755939\",\"4_90757272\",\"4_90757735\",\"4_90757840\",\"4_90757947\",\"4_90758945\")\n",
    "\n",
    "#Set binary pheno (0 ctrl, 1 pat):\n",
    "H4_SNCA$PHENO_01 <- H4_SNCA$PHENO-1\n",
    "y.bin <- 1*(H4_SNCA$PHENO_01==\"1\")\n",
    "\n",
    "#Non-adjusted:\n",
    "H4 <- haplo.cc(y=y.bin, geno=geno, locus.label= label, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H4, nlines=10, digits=2)\n",
    "\n",
    "#Sort the output on p-value:\n",
    "names(H4)\n",
    "H4_cc.df <- H4$cc.df\n",
    "H4_cc.df_sort <- H4_cc.df[order(H4_cc.df$`p-val`),]\n",
    "\n",
    "#Adjusted for age, sex, PC1-5:\n",
    "H4_adj <- haplo.cc(y=y.bin, geno=geno, locus.label= label, x.adj=adj, control = haplo.glm.control(haplo.freq.min = 0.01))\n",
    "print(H4_adj, nlines=10, digits=2)\n",
    "H4_adj_cc.df <- H4_adj$cc.df\n",
    "H4_adj_cc.df_sort <- H4_adj_cc.df[order(H4_adj_cc.df$`p-val`),]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you might notice, none of the haplotypes in none of the haplotype blocks was significantly associated with PD here. Not so surprising though since the cases and controls where randomized in this data :)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Haplotype analysis - Comparing the size of haplotype blocks at PD risk loci across different populations\n",
    "<a id=\"3\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* As an example, we will use the 1000G YRI and ACB populations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Extract cases from both the 1000G YRI and ACB populations:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/ACB_ALL_1000 --maf 0.01 --filter-cases --allow-no-sex --make-bed --out /home/jupyter/1000G_TEST/CASES.0.01.ACB\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/CASES.0.01.ACB.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /home/jupyter/1000G_TEST/ACB_ALL_1000\n",
      "  --filter-cases\n",
      "  --maf 0.01\n",
      "  --make-bed\n",
      "  --out /home/jupyter/1000G_TEST/CASES.0.01.ACB\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "34580269 variants loaded from .bim file.\n",
      "96 people (0 males, 0 females, 96 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to /home/jupyter/1000G_TEST/CASES.0.01.ACB.nosex .\n",
      "96 phenotype values loaded from .fam.\n",
      "1 person removed due to case/control status (--filter-cases).\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 95 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "18633487 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "15946782 variants and 95 people pass filters and QC.\n",
      "Among remaining phenotypes, 95 are cases and 0 are controls.\n",
      "--make-bed to /home/jupyter/1000G_TEST/CASES.0.01.ACB.bed +\n",
      "/home/jupyter/1000G_TEST/CASES.0.01.ACB.bim +\n",
      "/home/jupyter/1000G_TEST/CASES.0.01.ACB.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/ACB_ALL_1000 \\\n",
    "--maf 0.01 \\\n",
    "--filter-cases \\\n",
    "--allow-no-sex \\\n",
    "--make-bed \\\n",
    "--out /home/jupyter/1000G_TEST/CASES.0.01.ACB')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/YRI_ALL_1000g --maf 0.01 --filter-cases --allow-no-sex --make-bed --out /home/jupyter/1000G_TEST/CASES.0.01.YRI\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/CASES.0.01.YRI.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /home/jupyter/1000G_TEST/YRI_ALL_1000g\n",
      "  --filter-cases\n",
      "  --maf 0.01\n",
      "  --make-bed\n",
      "  --out /home/jupyter/1000G_TEST/CASES.0.01.YRI\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "34580269 variants loaded from .bim file.\n",
      "108 people (0 males, 0 females, 108 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to /home/jupyter/1000G_TEST/CASES.0.01.YRI.nosex .\n",
      "108 phenotype values loaded from .fam.\n",
      "0 people removed due to case/control status (--filter-cases).\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 108 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "20538652 variants removed due to minor allele threshold(s)\n",
      "(--maf/--max-maf/--mac/--max-mac).\n",
      "14041617 variants and 108 people pass filters and QC.\n",
      "Among remaining phenotypes, 108 are cases and 0 are controls.\n",
      "--make-bed to /home/jupyter/1000G_TEST/CASES.0.01.YRI.bed +\n",
      "/home/jupyter/1000G_TEST/CASES.0.01.YRI.bim +\n",
      "/home/jupyter/1000G_TEST/CASES.0.01.YRI.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/YRI_ALL_1000g \\\n",
    "--maf 0.01 \\\n",
    "--filter-cases \\\n",
    "--allow-no-sex \\\n",
    "--make-bed \\\n",
    "--out /home/jupyter/1000G_TEST/CASES.0.01.YRI')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As you can see, only one individual in the ACB dataset was a control (good for us!)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Extract common SNPs between both datasets:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/CASES.0.01.ACB --bmerge /home/jupyter/1000G_TEST/CASES.0.01.YRI --allow-no-sex --make-bed --out /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /home/jupyter/1000G_TEST/CASES.0.01.ACB\n",
      "  --bmerge /home/jupyter/1000G_TEST/CASES.0.01.YRI\n",
      "  --make-bed\n",
      "  --out /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "95 people loaded from /home/jupyter/1000G_TEST/CASES.0.01.ACB.fam.\n",
      "108 people to be merged from /home/jupyter/1000G_TEST/CASES.0.01.YRI.fam.\n",
      "Of these, 108 are new, while 0 are present in the base dataset.\n",
      "15946782 markers loaded from /home/jupyter/1000G_TEST/CASES.0.01.ACB.bim.\n",
      "14041617 markers to be merged from /home/jupyter/1000G_TEST/CASES.0.01.YRI.bim.\n",
      "Of these, 1030818 are new, while 13010799 are present in the base dataset.\n",
      "Warning: Variants 'rs547346964' and '.' have the same position.\n",
      "Performing single-pass merge (203 people, 16977600 variants).\n",
      "Merged fileset written to                     \n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged-merge.bed +\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged-merge.bim +\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged-merge.fam .\n",
      "16977600 variants loaded from .bim file.\n",
      "203 people (0 males, 0 females, 203 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged.nosex .\n",
      "203 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 203 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.879582.\n",
      "16977600 variants and 203 people pass filters and QC.\n",
      "Among remaining phenotypes, 203 are cases and 0 are controls.\n",
      "--make-bed to /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged.bed +\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged.bim +\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/CASES.0.01.ACB \\\n",
    "--bmerge /home/jupyter/1000G_TEST/CASES.0.01.YRI \\\n",
    "--allow-no-sex \\\n",
    "--make-bed \\\n",
    "--out /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Set a maf of 1% and exclude regions to speed up the coming haplotype block analyses. For this, import a file with regions to exclude:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/regions_exclude_haplotypes_size.txt /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/regions_exclude_haplotypes_size.txt...\n",
      "/ [1/1 files][   1010 B/   1010 B] 100% Done                                    \n",
      "Operation completed over 1 objects/1010.0 B.                                     \n"
     ]
    }
   ],
   "source": [
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/regions_exclude_haplotypes_size.txt /home/jupyter/1000G_TEST/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged --geno 0.01 --make-bed --autosome --exclude range /home/jupyter/1000G_TEST/regions_exclude_haplotypes_size.txt --out /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01.log.\n",
      "Options in effect:\n",
      "  --autosome\n",
      "  --bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged\n",
      "  --exclude range /home/jupyter/1000G_TEST/regions_exclude_haplotypes_size.txt\n",
      "  --geno 0.01\n",
      "  --make-bed\n",
      "  --out /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "16977600 variants loaded from .bim file.\n",
      "203 people (0 males, 0 females, 203 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01.nosex .\n",
      "203 phenotype values loaded from .fam.\n",
      "--exclude range: 7892503 variants excluded.\n",
      "--exclude range: 9085097 variants remaining.\n",
      "Warning: Ignoring phenotypes of missing-sex samples.  If you don't want those\n",
      "phenotypes to be ignored, use the --allow-no-sex flag.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 203 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "Total genotyping rate is 0.877901.\n",
      "2151875 variants removed due to missing genotype data (--geno).\n",
      "6933222 variants and 203 people pass filters and QC.\n",
      "Among remaining phenotypes, 203 are cases and 0 are controls.\n",
      "--make-bed to\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01.bed +\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01.bim +\n",
      "/home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged \\\n",
    "--geno 0.01 \\\n",
    "--make-bed \\\n",
    "--autosome \\\n",
    "--exclude range /home/jupyter/1000G_TEST/regions_exclude_haplotypes_size.txt \\\n",
    "--out /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01 --keep /home/jupyter/1000G_TEST/CASES.0.01.ACB.fam --make-bed --allow-no-sex --out /home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01\n",
      "  --keep /home/jupyter/1000G_TEST/CASES.0.01.ACB.fam\n",
      "  --make-bed\n",
      "  --out /home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "6933222 variants loaded from .bim file.\n",
      "203 people (0 males, 0 females, 203 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to /home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO.nosex .\n",
      "203 phenotype values loaded from .fam.\n",
      "--keep: 95 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 95 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "6933222 variants and 95 people pass filters and QC.\n",
      "Among remaining phenotypes, 95 are cases and 0 are controls.\n",
      "--make-bed to /home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO.bed +\n",
      "/home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO.bim +\n",
      "/home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01 \\\n",
    "--keep /home/jupyter/1000G_TEST/CASES.0.01.ACB.fam \\\n",
    "--make-bed \\\n",
    "--allow-no-sex \\\n",
    "--out /home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01 --keep /home/jupyter/1000G_TEST/CASES.0.01.YRI.fam --make-bed --allow-no-sex --out /home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO.log.\n",
      "Options in effect:\n",
      "  --allow-no-sex\n",
      "  --bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01\n",
      "  --keep /home/jupyter/1000G_TEST/CASES.0.01.YRI.fam\n",
      "  --make-bed\n",
      "  --out /home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "6933222 variants loaded from .bim file.\n",
      "203 people (0 males, 0 females, 203 ambiguous) loaded from .fam.\n",
      "Ambiguous sex IDs written to /home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO.nosex .\n",
      "203 phenotype values loaded from .fam.\n",
      "--keep: 108 people remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 108 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "6933222 variants and 108 people pass filters and QC.\n",
      "Among remaining phenotypes, 108 are cases and 0 are controls.\n",
      "--make-bed to /home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO.bed +\n",
      "/home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO.bim +\n",
      "/home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO.fam ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/variants.CASES_ACB_YRI_merged_geno0.01 \\\n",
    "--keep /home/jupyter/1000G_TEST/CASES.0.01.YRI.fam \\\n",
    "--make-bed \\\n",
    "--allow-no-sex \\\n",
    "--out /home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Calculate haplotype blocks in both cohorts**\n",
    "\n",
    "This step is usually done for all the genotype data. However, sicne this takes a lot of time, we previously excludes some regions in order for the analyses to run faster (it still takes a couple of hours though)\n",
    "\n",
    "***If you do not want to run this stepand save some time, there are output files uploaded for you. You can retreive the files running the cell below:***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_blocks.* /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_blocks.blocks...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_blocks.blocks.det...   \n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/ACB_blocks.log...          \n",
      "\\ [3/3 files][ 86.5 MiB/ 86.5 MiB] 100% Done                                    \n",
      "Operation completed over 3 objects/86.5 MiB.                                     \n",
      "Executing: gsutil -mu terra-9b559320 cp gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_blocks.* /home/jupyter/1000G_TEST/\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_blocks.blocks...\n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_blocks.blocks.det...   \n",
      "Copying gs://fc-c04486b2-8d7e-4359-a607-63643e9a7914/YRI_blocks.log...          \n",
      "\\ [3/3 files][ 90.2 MiB/ 90.2 MiB] 100% Done                                    \n",
      "Operation completed over 3 objects/90.2 MiB.                                     \n"
     ]
    }
   ],
   "source": [
    "#Retrieve output haplotype files for the ACB population:\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/ACB_blocks.* /home/jupyter/1000G_TEST/')\n",
    "\n",
    "#Retrieve output haplotype files for the YRI population:\n",
    "shell_do(f'gsutil -mu {BILLING_PROJECT_ID} cp {WORKSPACE_BUCKET}/YRI_blocks.* /home/jupyter/1000G_TEST/')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***If you instead prefer to run the commands yourself, run the following two cells. Otherwise, skip these steps and use the uploaded files!***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/CASES_ACB_TOHAPLO \\\n",
    "--blocks \\\n",
    "--allow-no-sex \\\n",
    "--out /home/jupyter/1000G_TEST/ACB_blocks')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/1000G_TEST/CASES_YRI_TOHAPLO \\\n",
    "--blocks \\\n",
    "--allow-no-sex \\\n",
    "--out /home/jupyter/1000G_TEST/YRI_blocks')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Compare the length and number of SNPs of haplotypes at the 92 risk loci between the two populations:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#Read the haplotype files:\n",
    "ACB_Haplo <- read.csv(\"/home/jupyter/1000G_TEST/ACB_blocks.blocks.det\", sep=\"\")\n",
    "YRI_Haplo <- read.csv(\"/home/jupyter/1000G_TEST/YRI_blocks.blocks.det\", sep=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "#PD loci file:\n",
    "PD_loci <- read.delim(\"/home/jupyter/1000G_TEST/92SNPs_haplotype.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " CHR          BP1          BP2           KB  NSNPS SNPS\n",
      "   1    154398961    154399353        0.393      2 rs111266048|rs78186112\n",
      "   1    154400015    154400320        0.306      2 rs4845618|rs6687726\n",
      "   1    154400600    154409719         9.12     32 rs80219185|rs6427658|rs73014271|rs7537306|rs58281895|rs79150137|rs28730665|rs2228144|rs2229237|rs6694817|rs28730733|rs73014279|rs74709806|rs7549250|rs7549338|rs7553796|rs114766856|rs56383622|rs4845619|rs114645658|rs77676154|rs78307575|rs59632925|rs4845620|rs7518199|rs7521458|rs115321795|rs4845371|rs57502626|rs6667434|rs55826755|rs149274188\n",
      "   1    154409814    154410481        0.668      2 rs181088116|rs113588579\n",
      "   1    154410482    154410490        0.009      2 rs61812592|rs61812593\n",
      "   1    154411419    154416969        5.551     15 rs4845622|rs59502179|rs4393147|rs4453032|rs6664201|rs61812596|rs115093244|rs201335322|rs7529670|rs4845372|rs4845623|rs6676117|rs182632843|rs12753254|rs12730036\n",
      "   1    154418749    154422067        3.319     17 rs4576655|rs4537545|rs11265614|rs58229059|rs111301013|rs6686750|rs12730935|rs558204390|rs61812598|rs7512646|rs7529229|rs12404927|rs12403159|rs12404936|rs61812599|rs111808975|rs4845625\n",
      "   1    154422781    154423731        0.951      3 rs149402993|rs4845626|rs111685209\n",
      "   1    154423764    154423909        0.146      2 rs12129500|rs7536152\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head -10 /home/jupyter/1000G_TEST/ACB_blocks.blocks.det"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "PD_loci$ACB.HapLength.KB <- NA\n",
    "PD_loci$ACB.HapSnps.N <- NA\n",
    "PD_loci$ACB.HapSnps.RS <- NA\n",
    "PD_loci$YRI.HapLength.KB <- NA\n",
    "PD_loci$YRI.HapSnps.N <- NA\n",
    "PD_loci$YRI.HapSnps.RS <- NA\n",
    "for(i in 1:length(PD_loci$SNP))\n",
    "{\n",
    "  thisSnp <- PD_loci$SNP[i]\n",
    "  thisChr <- PD_loci$CHR[i]\n",
    "  thisBp <- PD_loci$BP[i]\n",
    "  ACBHap <- subset(ACB_Haplo, CHR == thisChr & BP1 <= thisBp & BP2 >= thisBp)\n",
    "  YRIHap <- subset(YRI_Haplo, CHR == thisChr & BP1 <= thisBp & BP2 >= thisBp)\n",
    "  if(length(ACBHap$KB) > 0) \n",
    "  {\n",
    "    PD_loci$ACB.HapLength.KB[i] <- ACBHap$KB\n",
    "    PD_loci$ACB.HapSnps.N[i] <- ACBHap$NSNPS\n",
    "    PD_loci$ACB.HapSnps.RS[i] <- ACBHap$SNPS\n",
    "  }\n",
    "  if(length(YRIHap$KB) > 0) \n",
    "  {\n",
    "    PD_loci$YRI.HapLength.KB[i] <- YRIHap$KB\n",
    "    PD_loci$YRI.HapSnps.N[i] <- YRIHap$NSNPS\n",
    "    PD_loci$YRI.HapSnps.RS[i] <- YRIHap$SNPS\n",
    "  }\n",
    "}\n",
    "PD_loci_sort <- PD_loci[order(PD_loci$`ACB.HapSnps.N`),]\n",
    "fwrite(PD_loci_sort, \"/home/jupyter/1000G_TEST/PD_lociHaplo.tab\", quote = F, sep = \"\\t\", row.names = F, na = NA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP\tCHR\tBP\tACB.HapLength.KB\tACB.HapSnps.N\tACB.HapSnps.RS\tYRI.HapLength.KB\tYRI.HapSnps.N\tYRI.HapSnps.RS\n",
      "rs62333164\t4\t170583157\t0.514\t2\trs72694777|rs62333164\t0.514\t2\trs72694777|rs62333164\n",
      "rs11950533\t5\t134199105\t0.673\t2\trs73282857|rs111519842\t0.673\t2\trs73282857|rs111519842\n",
      "rs3802920\t11\t133787001\t0.009\t2\trs3802921|rs3802920\t1.201\t6\trs75125504|rs3824992|rs3802923|rs3802922|rs75952732|rs3741104\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "head -4 /home/jupyter/1000G_TEST/PD_lociHaplo.tab"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The output file PD_lociHaplo.tab contain the lenght of the haplotype block in KB (HapLength.KB), the number of SNPs in the block (HapSnps.N) and the rsID of the SNPs for both populations mapped to the PD risk loci (SNPs) in the input file 92SNPs_haplotype.txt. This makes in possible for us to compare the haplotype blocks in the different populations at the risk loci. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Rare variant analyses - Gene/Region-based analysis (including annotation using ANNOVAR)\n",
    "<a id=\"4\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* We will focus on a specific region of interest and as an example, we will again look at the region of the gene SNCA\n",
    "\n",
    "* SNCA position on GRCh37/hg19:chr4:90645250-90759466 (Ensembl)\n",
    "\n",
    "* First, we create a new folder for the analyses"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p /home/jupyter/Burden/SNCA_burden"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Extract the region of interest**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted --chr 4 --from-bp 90645250 --to-bp 90759466 --make-bed --out /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted\n",
      "  --chr 4\n",
      "  --from-bp 90645250\n",
      "  --make-bed\n",
      "  --out /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA\n",
      "  --to-bp 90759466\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "600 out of 12405097 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "600 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--make-bed to\n",
      "/home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.bed +\n",
      "/home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.bim +\n",
      "/home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.fam ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted \\\n",
    "--chr 4 \\\n",
    "--from-bp 90645250 \\\n",
    "--to-bp 90759466 \\\n",
    "--make-bed \\\n",
    "--out /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Recode the PLINK files to vcf**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA --recode vcf-fid --out /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA\n",
      "  --out /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA\n",
      "  --recode vcf-fid\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "600 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "600 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--recode vcf-fid to\n",
      "/home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.vcf ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA \\\n",
    "--recode vcf-fid \\\n",
    "--out /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**bgzip and tabix the vcf file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "export PATH=$PATH:/home/jupyter/rvtests/third/tabix-0.2.6/\n",
    "bgzip /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.vcf\n",
    "tabix -f -p vcf /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Annotate the vcf file using ANNOVAR**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%bash\n",
    "\n",
    "perl /home/jupyter/annovar/table_annovar.pl /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA.vcf.gz /home/jupyter/annovar/humandb/ -buildver hg19 \\\n",
    "-out /home/jupyter/Burden/SNCA_burden/SNCA.annovar \\\n",
    "-remove -protocol refGene,ljb26_all,gnomad211_genome,clinvar_20140902 \\\n",
    "-operation g,f,f,f \\\n",
    "-nastring . \\\n",
    "-vcfinput"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Clean up the output file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "head -1 /home/jupyter/Burden/SNCA_burden/SNCA.annovar.hg19_multianno.txt > /home/jupyter/Burden/SNCA_burden/header.txt\n",
    "colct=\"$(wc -w /home/jupyter/Burden/SNCA_burden/header.txt| cut -f1 -d' ')\"\n",
    "cut -f1-$colct /home/jupyter/Burden/SNCA_burden/SNCA.annovar.hg19_multianno.txt > /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Generate list of non-coding variants**\n",
    "\n",
    "The following can also be done for all variants or other functional categories such as non-coding or CADD. (CADD = scoring of the deleteriousness variants). However, here we will look at the non-coding variants as an example (There is only one coding variant present in this dataset, hence we look at the noncodinh variants)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "awk '$6==\"intronic\" {print}' /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.txt > /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.txt\n",
    "awk '{print $1\" \"$2\" \"$2\" \"$7}' /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.txt > /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Below, we will have a look at the non-coding variants in the region in our dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4 90647865 90647865 SNCA\n",
      "4 90648185 90648185 SNCA\n",
      "4 90648482 90648482 SNCA\n",
      "4 90648686 90648686 SNCA\n",
      "4 90648952 90648952 SNCA\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "head -5 /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Extract the non-conding variants from the data:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA --extract range /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt --recode vcf-fid --out /home/jupyter/Burden/SNCA_burden/SNCA.noncoding\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA\n",
      "  --extract range /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt\n",
      "  --out /home/jupyter/Burden/SNCA_burden/SNCA.noncoding\n",
      "  --recode vcf-fid\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "600 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "--extract range: 19 variants excluded.\n",
      "--extract range: 581 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "581 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--recode vcf-fid to /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.vcf ...\n",
      "101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA \\\n",
    "--extract range /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt \\\n",
    "--recode vcf-fid \\\n",
    "--out /home/jupyter/Burden/SNCA_burden/SNCA.noncoding')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Recode to generate a .ped file with phenotype info to use in RVTESTS**\n",
    "\n",
    "The 6th column of the phenotype file .ped, which is in PLINK format, will be used. Rvtests will automatically check whether the phenotype is binary trait or quantitative trait. For binary trait, the recommended way of coding is to code controls as 1, cases as 2, missing phenotypes as -9 or 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA --extract range /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt --recode --out /home/jupyter/Burden/SNCA_burden/SNCA.noncoding\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA\n",
      "  --extract range /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt\n",
      "  --out /home/jupyter/Burden/SNCA_burden/SNCA.noncoding\n",
      "  --recode\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "600 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "--extract range: 19 variants excluded.\n",
      "--extract range: 581 variants remaining.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "581 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--recode ped to /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.ped +\n",
      "/home/jupyter/Burden/SNCA_burden/SNCA.noncoding.map ... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/Burden/SNCA_burden/IMPUTED.HARDCALLS.Demo_formatted.SNCA \\\n",
    "--extract range /home/jupyter/Burden/SNCA_burden/SNCA.trimmed.annotation.noncoding.variants.SNPs.txt \\\n",
    "--recode \\\n",
    "--out /home/jupyter/Burden/SNCA_burden/SNCA.noncoding')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Run rare variant analyses in RVTESTS**\n",
    "\n",
    "RVTESTS is great since we can run all the test we are interested in at the same time:\n",
    "* Burden test (cmc,zeggini,mb,fp,cmcWald)\n",
    "* Non-burden sequence kernel association test (SKAT)\n",
    "* Optimal SKAT (SKAT-O)\n",
    "\n",
    "cmc,zeggini,mb,fp,cmcWald are all burden analyses but different ways to calculate/different algorithms. The website for RTVTESTS has nice explainations for all analyses and references if you are interested in learning more: http://zhanxw.github.io/rvtests/#burden-tests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Bgzip and index the vcf file**\n",
    "\n",
    "RVTESTS supports vcf files, and files in both plain text format or gzipped format are supported. But to use group-based rare variant tests (which we are going to), indexed the VCF files using tabix are required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "export PATH=$PATH:/home/jupyter/rvtests/third/tabix-0.2.6/\n",
    "bgzip /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.vcf\n",
    "tabix -f -p vcf /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.vcf.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Run the analysis**\n",
    "Here we use an upper minor allele frequency of 1% (--freqUpper 0.01)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "%%bash\n",
    "export PATH=$PATH:/home/jupyter/rvtests/executable\n",
    "\n",
    "rvtest --noweb \\\n",
    "--inVcf /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.vcf.gz \\\n",
    "--pheno /home/jupyter/Burden/SNCA_burden/SNCA.noncoding.ped \\\n",
    "--covar /home/jupyter/Burden/covariateFile_Haplo.txt \\\n",
    "--covar-name SEX,AGE,PC1,PC2,PC3,PC4,PC5 \\\n",
    "--kernel skat,skato \\\n",
    "--burden cmc,zeggini,mb,fp,cmcWald \\\n",
    "--geneFile /home/jupyter/Burden/refFlat_hg19.txt \\\n",
    "--freqUpper 0.01 \\\n",
    "--out /home/jupyter/Burden/SNCA_burden/BURDEN.SNCA\n",
    "\n",
    "# --out : Name of output \n",
    "# --burden X --kernel X: tests to run \n",
    "# --inVcf : path to VCF file \n",
    "# --pheno : path to pheno+covariate file\n",
    "# --pheno-name : column name with phenotype in file\n",
    "# --covar : path to pheno+covariate file\n",
    "# --freqUpper : MAF cut-off\n",
    "# --covar-name : covariates, listed by column name, separated by commas (no spaces between commas)\n",
    "# --geneFile : refFlat path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BURDEN.SNCA.CMC.assoc\n",
      "BURDEN.SNCA.CMCWald.assoc\n",
      "BURDEN.SNCA.Fp.assoc\n",
      "BURDEN.SNCA.log\n",
      "BURDEN.SNCA.MadsonBrowning.assoc\n",
      "BURDEN.SNCA.Skat.assoc\n",
      "BURDEN.SNCA.SkatO.assoc\n",
      "BURDEN.SNCA.Zeggini.assoc\n",
      "header.txt\n",
      "IMPUTED.HARDCALLS.Demo_formatted.SNCA.bed\n",
      "IMPUTED.HARDCALLS.Demo_formatted.SNCA.bim\n",
      "IMPUTED.HARDCALLS.Demo_formatted.SNCA.fam\n",
      "IMPUTED.HARDCALLS.Demo_formatted.SNCA.log\n",
      "IMPUTED.HARDCALLS.Demo_formatted.SNCA.vcf.gz\n",
      "IMPUTED.HARDCALLS.Demo_formatted.SNCA.vcf.gz.tbi\n",
      "SNCA.annovar.avinput\n",
      "SNCA.annovar.hg19_multianno.txt\n",
      "SNCA.annovar.hg19_multianno.vcf\n",
      "SNCA.noncoding.log\n",
      "SNCA.noncoding.map\n",
      "SNCA.noncoding.ped\n",
      "SNCA.noncoding.vcf.gz\n",
      "SNCA.noncoding.vcf.gz.tbi\n",
      "SNCA.trimmed.annotation.noncoding.variants.SNPs.txt\n",
      "SNCA.trimmed.annotation.noncoding.variants.txt\n",
      "SNCA.trimmed.annotation.txt\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/jupyter/Burden/SNCA_burden/\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene\tRANGE\tN_INFORMATIVE\tNumVar\tNumPolyVar\tQ\trho\tPvalue\n",
      "SNCA\t4:90645249-90758350,4:90645249-90758127,4:90645249-90758127,4:90645249-90759447\t404\t193\t193\t17325.8\t0.3\t0.366207\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/jupyter/Burden/SNCA_burden/\n",
    "head BURDEN.SNCA.SkatO.assoc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As an example, you can see the output file for the SKAT-O analysis. The p-value for the analysis is approximately 0.37, indicating that the cumulative effect of rare (MAF<1%) noncoding variants in SNCA in our data do are not associated with PD "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Rare variant analyses - Pathway-based analysis\n",
    "<a id=\"5\"></a>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Run gene-set based pathway test in RVESTs**\n",
    "\n",
    "For this, we need a SetFile which is a tab-delimited file with the following format:\n",
    "* Each pathway has it's own line, and gene coordinates follow in this format (no spaces between commas) 16:68245303-68261058,15:82659280-82709946 (each pathway is a set)\n",
    "* PATHWAYNAME1 gene1coordinates,gene2coordinates etc. E.g., KEGG_LYSOSOME 16:68245303-68261058,16:5024843-5034141,16:71728999-71809201, ....\n",
    "* PATHWAYNAME2 gene1coordinates,gene2coordinates etc."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Recode the PLINK format file to vcf and bgzip:**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Executing: /home/jupyter/plink --bfile /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted --recode vcf-fid --out /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted\n",
      "PLINK v1.90b6.9 64-bit (4 Mar 2019)            www.cog-genomics.org/plink/1.9/\n",
      "(C) 2005-2019 Shaun Purcell, Christopher Chang   GNU General Public License v3\n",
      "Logging to /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.log.\n",
      "Options in effect:\n",
      "  --bfile /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted\n",
      "  --out /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted\n",
      "  --recode vcf-fid\n",
      "\n",
      "15000 MB RAM detected; reserving 7500 MB for main workspace.\n",
      "12405097 variants loaded from .bim file.\n",
      "404 people (202 males, 202 females) loaded from .fam.\n",
      "404 phenotype values loaded from .fam.\n",
      "Using 1 thread (no multithreaded calculations invoked).\n",
      "Before main variant filters, 404 founders and 0 nonfounders present.\n",
      "Calculating allele frequencies... 10111213141516171819202122232425262728293031323334353637383940414243444546474849505152535455565758596061626364656667686970717273747576777879808182838485868788899091929394959697989 done.\n",
      "12405097 variants and 404 people pass filters and QC.\n",
      "Among remaining phenotypes, 202 are cases and 202 are controls.\n",
      "--recode vcf-fid to /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf\n",
      "... 101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899done.\n"
     ]
    }
   ],
   "source": [
    "shell_do(f'/home/jupyter/plink \\\n",
    "--bfile /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted \\\n",
    "--recode vcf-fid \\\n",
    "--out /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "export PATH=$PATH:/home/jupyter/rvtests/third/tabix-0.2.6/\n",
    "bgzip /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf\n",
    "tabix -f -p vcf /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "covariateFile_Haplo.txt\n",
      "IMPUTED.HARDCALLS.Demo_formatted.bed\n",
      "IMPUTED.HARDCALLS.Demo_formatted.bim\n",
      "IMPUTED.HARDCALLS.Demo_formatted.fam\n",
      "IMPUTED.HARDCALLS.Demo_formatted.log\n",
      "IMPUTED.HARDCALLS.Demo_formatted.vcf.gz\n",
      "IMPUTED.HARDCALLS.Demo_formatted.vcf.gz.tbi\n",
      "lysosomal_burden.log\n",
      "lysosomal-setfile-example.txt\n",
      "refFlat_hg19.txt\n",
      "SNCA_burden\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/jupyter/Burden/\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Run the pathway-based burden analysis in RVTESTS**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thank you for using rvtests (version: 20190205, git: 8defd6fcbba91ae5187ea8fbce6ccc9c944bb4cb)\n",
      "  For documentations, refer to http://zhanxw.github.io/rvtests/\n",
      "  For questions and comments, plase send to Xiaowei Zhan <zhanxw@umich.edu>\n",
      "  For bugs and feature requests, please submit at: https://github.com/zhanxw/rvtests/issues\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "The following parameters are available.  Ones with \"[]\" are in effect:\n",
      "\n",
      "Available Options\n",
      "      Basic Input/Output:\n",
      "                          --inVcf [/home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf.gz]\n",
      "                          --inBgen [], --inBgenSample [], --inKgg []\n",
      "                          --out [/home/jupyter/Burden/lysosomal_burden]\n",
      "                          --outputRaw\n",
      "       Specify Covariate: --covar [/home/jupyter/Burden/covariateFile_Haplo.txt]\n",
      "                          --covar-name [SEX,AGE,PC1,PC2,PC3,PC4,PC5], --sex\n",
      "       Specify Phenotype: --pheno [/home/jupyter/Burden/covariateFile_Haplo.txt]\n",
      "                          --inverseNormal, --useResidualAsPhenotype, --mpheno []\n",
      "                          --pheno-name [PHENO], --qtl, --multiplePheno []\n",
      "        Specify Genotype: --dosage [], --multipleAllele\n",
      "    Chromosome X Options: --xLabel [], --xParRegion []\n",
      "           People Filter: --peopleIncludeID [], --peopleIncludeFile []\n",
      "                          --peopleExcludeID [], --peopleExcludeFile []\n",
      "             Site Filter: --rangeList [], --rangeFile [], --siteFile []\n",
      "                          --siteDepthMin [], --siteDepthMax [], --siteMACMin []\n",
      "                          --annoType []\n",
      "         Genotype Filter: --indvDepthMin [], --indvDepthMax [], --indvQualMin []\n",
      "       Association Model: --single [], --burden [], --vt [], --kernel []\n",
      "                          --meta []\n",
      "     Family-based Models: --kinship [], --xHemiKinship [], --kinshipEigen []\n",
      "                          --xHemiKinshipEigen [], --boltPlink []\n",
      "                          --boltPlinkNoCheck\n",
      "          Grouping Unit : --geneFile [], --gene [], --setList []\n",
      "                          --setFile [/home/jupyter/Burden/lysosomal-setfile-example.txt]\n",
      "                          --set []\n",
      "        Frequency Cutoff: --freqUpper [0.01], --freqLower []\n",
      "            Missing Data: --impute [], --imputePheno, --imputeCov\n",
      "    Conditional Analysis: --condition []\n",
      "     Auxiliary Functions: --noweb [true], --hide-covar [true], --numThread []\n",
      "                          --outputID, --help\n",
      "\n",
      "Effective Options\n",
      "    --inVcf /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf.gz\n",
      "    --out /home/jupyter/Burden/lysosomal_burden\n",
      "    --covar /home/jupyter/Burden/covariateFile_Haplo.txt\n",
      "    --covar-name SEX,AGE,PC1,PC2,PC3,PC4,PC5\n",
      "    --pheno /home/jupyter/Burden/covariateFile_Haplo.txt\n",
      "    --pheno-name PHENO\n",
      "    --setFile /home/jupyter/Burden/lysosomal-setfile-example.txt\n",
      "    --freqUpper 0.01\n",
      "    --noweb\n",
      "    --hide-covar\n",
      "\n",
      "Unparsed arguments:  --bruden cmc"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'export PATH=$PATH:/home/jupyter/rvtests/executable\\n\\nrvtest --noweb --hide-covar \\\\\\n--out /home/jupyter/Burden/lysosomal_burden \\\\\\n--bruden cmc \\\\\\n--inVcf /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf.gz \\\\\\n--pheno /home/jupyter/Burden/covariateFile_Haplo.txt \\\\\\n--pheno-name PHENO \\\\\\n--covar /home/jupyter/Burden/covariateFile_Haplo.txt \\\\\\n--covar-name SEX,AGE,PC1,PC2,PC3,PC4,PC5 \\\\\\n--freqUpper 0.01 \\\\\\n--setFile /home/jupyter/Burden/lysosomal-setfile-example.txt\\n\\n# --out : Name of output \\n# --burden cmc --kernel skato: tests to run \\n# --inVcf : path to VCF file \\n# --pheno : path to pheno+covariate file\\n# --pheno-name : column name with phenotype in file\\n# --covar : path to pheno+covariate file\\n# --freqUpper : optional, MAF cut-off\\n# --covar-name : covariates, listed by column name, separated by commas (no spaces between commas)\\n# --set : optional, run individual set \\n# --setFile : setFile path\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_1456/3594953804.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'bash'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'export PATH=$PATH:/home/jupyter/rvtests/executable\\n\\nrvtest --noweb --hide-covar \\\\\\n--out /home/jupyter/Burden/lysosomal_burden \\\\\\n--bruden cmc \\\\\\n--inVcf /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf.gz \\\\\\n--pheno /home/jupyter/Burden/covariateFile_Haplo.txt \\\\\\n--pheno-name PHENO \\\\\\n--covar /home/jupyter/Burden/covariateFile_Haplo.txt \\\\\\n--covar-name SEX,AGE,PC1,PC2,PC3,PC4,PC5 \\\\\\n--freqUpper 0.01 \\\\\\n--setFile /home/jupyter/Burden/lysosomal-setfile-example.txt\\n\\n# --out : Name of output \\n# --burden cmc --kernel skato: tests to run \\n# --inVcf : path to VCF file \\n# --pheno : path to pheno+covariate file\\n# --pheno-name : column name with phenotype in file\\n# --covar : path to pheno+covariate file\\n# --freqUpper : optional, MAF cut-off\\n# --covar-name : covariates, listed by column name, separated by commas (no spaces between commas)\\n# --set : optional, run individual set \\n# --setFile : setFile path\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2404\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2405\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2406\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2407\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2408\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mnamed_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m                 \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscript\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshebang\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0;31m# write a basic docstring:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/decorator.py\u001b[0m in \u001b[0;36mfun\u001b[0;34m(*args, **kw)\u001b[0m\n\u001b[1;32m    230\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mkwsyntax\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    231\u001b[0m                 \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkw\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msig\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 232\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mcaller\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mextras\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    233\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__name__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m     \u001b[0mfun\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__doc__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/opt/conda/lib/python3.7/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    243\u001b[0m             \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_error\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 245\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_run_script\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_close\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'export PATH=$PATH:/home/jupyter/rvtests/executable\\n\\nrvtest --noweb --hide-covar \\\\\\n--out /home/jupyter/Burden/lysosomal_burden \\\\\\n--bruden cmc \\\\\\n--inVcf /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf.gz \\\\\\n--pheno /home/jupyter/Burden/covariateFile_Haplo.txt \\\\\\n--pheno-name PHENO \\\\\\n--covar /home/jupyter/Burden/covariateFile_Haplo.txt \\\\\\n--covar-name SEX,AGE,PC1,PC2,PC3,PC4,PC5 \\\\\\n--freqUpper 0.01 \\\\\\n--setFile /home/jupyter/Burden/lysosomal-setfile-example.txt\\n\\n# --out : Name of output \\n# --burden cmc --kernel skato: tests to run \\n# --inVcf : path to VCF file \\n# --pheno : path to pheno+covariate file\\n# --pheno-name : column name with phenotype in file\\n# --covar : path to pheno+covariate file\\n# --freqUpper : optional, MAF cut-off\\n# --covar-name : covariates, listed by column name, separated by commas (no spaces between commas)\\n# --set : optional, run individual set \\n# --setFile : setFile path\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "export PATH=$PATH:/home/jupyter/rvtests/executable\n",
    "\n",
    "rvtest --noweb --hide-covar \\\n",
    "--out /home/jupyter/Burden/lysosomal_burden \\\n",
    "--bruden cmc \\\n",
    "--inVcf /home/jupyter/Burden/IMPUTED.HARDCALLS.Demo_formatted.vcf.gz \\\n",
    "--pheno /home/jupyter/Burden/covariateFile_Haplo.txt \\\n",
    "--pheno-name PHENO \\\n",
    "--covar /home/jupyter/Burden/covariateFile_Haplo.txt \\\n",
    "--covar-name SEX,AGE,PC1,PC2,PC3,PC4,PC5 \\\n",
    "--freqUpper 0.01 \\\n",
    "--setFile /home/jupyter/Burden/lysosomal-setfile-example.txt\n",
    "\n",
    "# --out : Name of output \n",
    "# --burden cmc --kernel skato: tests to run \n",
    "# --inVcf : path to VCF file \n",
    "# --pheno : path to pheno+covariate file\n",
    "# --pheno-name : column name with phenotype in file\n",
    "# --covar : path to pheno+covariate file\n",
    "# --freqUpper : optional, MAF cut-off\n",
    "# --covar-name : covariates, listed by column name, separated by commas (no spaces between commas)\n",
    "# --set : optional, run individual set \n",
    "# --setFile : setFile path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Range\tRANGE\tN_INFORMATIVE\tNumVar\tNumPolyVar\tNonRefSite\tPvalue\n",
      "KEGG_LYSOSOME\t16:68245303-68261058,16:5024843-5034141,16:71728999-71809201,16:23463541-23521995,16:88813733-88856970,16:2513951-2520218,16:1351930-1364113,16:67438013-67481181,16:28474110-28495575,15:82659280-82709946,15:78921057-78949574,15:50908671-51005895,15:89830598-89894638,15:72340923-72376420,22:37608724-37633564,22:19179472-19291719,22:29327679-29423179,22:50622753-50628173,22:42058333-42070842,2:218382028-218396894,2:223751685-223838027,2:20032649-20051628,1:40071460-40097727,1:43974486-43978295,1:150730187-150765792,1:30732468-30757774,1:113894193-113905201,1:150796207-150808260,1:206009263-206023895,1:155234451-155244699,1:84398531-84415018,1:109309567-109397918,1:23845076-23868294,4:127917798-127966034,4:155924117-155953866,4:177430773-177442437,4:7430284-7434930,4:986996-1004564,4:102630769-102760994,4:76158736-76234536,20:58995184-59007254,20:45890143-45898820,6:159969098-160113507,6:109366513-109382467,6:32153440-32163678,6:31857658-31862822,6:73593378-73653992,17:68259181-68422731,17:42458843-42522611,17:75236598-75262363,17:80206715-80220923,17:7579490-7582111,17:3636458-3663103,17:59619688-59696956,17:80101555-80119881,17:42536240-42544449,11:65879836-65883741,11:6390439-6394998,11:88293591-88337761,11:1752754-1764573,11:68039024-68050895,11:66563463-66568841,11:6612767-6619448,11:47239301-47248906,14:92703806-92748679,14:31025105-31130996,14:24573517-24576250,14:87837819-87993665,14:74476191-74494177,X:120426147-120469365,X:154428644-154436516,X:149476987-149521096,X:101393272-101408012,X:15825805-15854931,5:78002325-78294698,5:151212149-151270440,5:78777208-78986087,5:74640022-74722647,5:115841591-115914081,5:176392454-176416569,3:183122214-183163839,3:32996608-33097202,3:3700813-4467273,3:50299889-50312381,7:138706293-138799560,7:100101548-100110345,7:65960683-65982215,7:101154455-101161596,9:87726108-87731469,9:97029676-97039643,9:137007233-137028922,9:36190855-36304781,12:122920950-122981649,12:50979400-51028566,12:101745498-101830959,12:64713444-64759431,12:8940360-8949761,12:123712352-123761755,12:55725322-55729707,13:113297238-113323672,13:76990659-77019143,10:74120254-74151063,10:71816297-71851325,10:89213568-89414557,19:2100987-2164468,19:12875208-12881466,19:10572670-10587315,19:12646510-12666742,19:16197577-16245907,19:50358476-50365830,19:7522623-7534009,19:11574659-11579008,8:53715542-53843558,8:43140463-43202855,8:42152945-42171673,8:85987322-86154225,8:11842523-11869448,8:18055991-18084998,8:23385782-23457695,8:97775056-97853013,18:23506183-23586506\t404\t10564\t10564\t404\t0.00262331\n",
      "REACTOME_LYSOSOME_VESICLE_BIOGENESIS\t12:8940360-8949761,12:55716036-55720087,12:64713444-64759431,19:16197577-16245907,19:10718078-10833488,19:12875208-12881466,19:10572670-10587315,19:58551565-58555105,15:50908671-51005895,22:29327679-29423179,14:31025105-31130996,14:23559564-23568070,20:58995184-59007254,7:101154455-101161596,7:100101548-100110345,9:17579065-17797124,9:36190855-36304781,11:123057488-123063230,11:75260121-75351705,1:65248218-65415871,1:113894193-113905201,1:228082659-228099212,2:85561561-85582031,2:223751685-223838027,X:155881344-155943769,X:15825805-15854931,17:59619688-59696956,17:81683325-81703138,17:8159148-8163546,21:25880549-26171128,6:122996234-123072925,6:7881516-7910788,16:71728999-71809201,5:176392454-176416569,8:61057157-61501645\t404\t5834\t5834\t404\t0.00262331\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "cd /home/jupyter/Burden/\n",
    "head lysosomal_burden.CMC.assoc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, the pathwaya appear to be significantly associated with PD (p=0.0026), but again, case-control status was randomized in this data. \n",
    "\n",
    "Even though this analysis can give us an indication regarding the association of pathway-specific genetic variation on PD risk, it is important to know that RVTESTS has some limitations for pathway analyses:\n",
    "* Rvtests relies on genomic coordinates, so prone to gene overlap/error\n",
    "* Rvtests doesn't take into account strand direction\n",
    "\n",
    "If you are interested to learn more about functional pathways in PD, please have a look at the paper by Bandres-Ciga et al: Large-scale pathway specific polygenic risk and transcriptomic community network analysis identifies novel functional pathways in Parkinson disease, 2020, doi: 10.1007/s00401-020-02181-3"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
